<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="AgueroZZ" />

<meta name="date" content="2024-09-13" />

<title>Circadian Detection: Simulation (Easy)</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">resultsummary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="simulation_pollution.html">Simulation 1</a>
</li>
<li>
  <a href="eQTL_sim.html">Simulation 2</a>
</li>
<li>
  <a href="covid_example.html">COVID-19 Example</a>
</li>
<li>
  <a href="expression.html">Gene Expression Example</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/AgueroZZ/resultsummary">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Circadian Detection: Simulation (Easy)</h1>
<h4 class="author">AgueroZZ</h4>
<h4 class="date">2024-09-13</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-09-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>FASHresultsummary/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240507code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240507)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240507code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240507)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreea16501d2d1af849b9cdaa1807aa316e413404981targetblanka16501da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/a16501d2d1af849b9cdaa1807aa316e413404981" target="_blank">a16501d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreea16501d2d1af849b9cdaa1807aa316e413404981targetblanka16501da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/a16501d2d1af849b9cdaa1807aa316e413404981" target="_blank">a16501d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    code/.DS_Store
    Ignored:    code/cpp/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/expression_data/.DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/example/.DS_Store
    Ignored:    output/example/figure/
    Ignored:    output/simulation_pollution/.DS_Store
    Ignored:    output/simulation_pollution/figure/

Unstaged changes:
    Modified:   analysis/circadian_sim.Rmd
    Deleted:    code/cpp/Poisson.o
    Deleted:    code/cpp/Poisson.so
    Deleted:    code/cpp/Poisson_expression.o
    Deleted:    code/cpp/Poisson_expression.so
    Deleted:    code/cpp/Poisson_just_fixed.o
    Deleted:    code/cpp/Poisson_just_fixed.so
    Deleted:    code/cpp/Poisson_just_fixed_expression.o
    Deleted:    code/cpp/Poisson_just_fixed_expression.so
    Modified:   thought_process

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/circadian_sim_easy.Rmd</code>)
and HTML (<code>docs/circadian_sim_easy.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/circadian_sim_easy.html" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/3573d41d4ffd5696ddafbe558f8b77fd4f0cfb1c/analysis/circadian_sim_easy.Rmd" target="_blank">3573d41</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
<td>
workflowr::wflow_publish("analysis/circadian_sim_easy.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(BayesGP)
library(TMB)
library(Matrix)
library(splines)
library(parallel)
library(ggplot2)
library(reshape2)
library(mixsqp)
library(tidyverse)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ lubridate 1.9.3     ✔ tibble    3.2.1
✔ purrr     1.0.2     ✔ tidyr     1.3.0
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ tidyr::expand() masks Matrix::expand()
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
✖ tidyr::pack()   masks Matrix::pack()
✖ tidyr::unpack() masks Matrix::unpack()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<pre class="r"><code>cpp_dir &lt;- paste0(getwd(), &quot;/code/cpp&quot;)
fig_dir &lt;- paste0(getwd(), &quot;/output/circadian_sim_easy&quot;)
result_dir &lt;- paste0(getwd(), &quot;/output/circadian_sim_easy&quot;)
function_dir &lt;- paste0(getwd(), &quot;/code/function&quot;)
source(paste0(function_dir, &quot;/functions_fitting_Poisson_circadian_expression.R&quot;))
source(paste0(function_dir, &quot;/functions_sim_Poisson_circadian_expression.R&quot;))
compile(paste0(cpp_dir, &quot;/Poisson_expression.cpp&quot;))</code></pre>
<pre><code>Warning in readLines(file): incomplete final line found on
&#39;/Users/ziangzhang/Desktop/FASH/FASHresultsummary/code/cpp/Poisson_expression.cpp&#39;</code></pre>
<pre><code>Note: Using Makevars in /Users/ziangzhang/.R/Makevars </code></pre>
<pre><code>using C++ compiler: &#39;Apple clang version 13.1.6 (clang-1316.0.21.2.5)&#39;
using SDK: &#39;MacOSX12.3.sdk&#39;</code></pre>
<pre><code>[1] 0</code></pre>
<pre class="r"><code>compile(paste0(cpp_dir, &quot;/Poisson_just_fixed_expression.cpp&quot;))</code></pre>
<pre><code>Warning in readLines(file): incomplete final line found on
&#39;/Users/ziangzhang/Desktop/FASH/FASHresultsummary/code/cpp/Poisson_just_fixed_expression.cpp&#39;</code></pre>
<pre><code>Note: Using Makevars in /Users/ziangzhang/.R/Makevars </code></pre>
<pre><code>using C++ compiler: &#39;Apple clang version 13.1.6 (clang-1316.0.21.2.5)&#39;
using SDK: &#39;MacOSX12.3.sdk&#39;</code></pre>
<pre><code>[1] 0</code></pre>
<pre class="r"><code>dyn.load(TMB::dynlib(paste0(cpp_dir, &quot;/Poisson_expression&quot;)))
dyn.load(TMB::dynlib(paste0(cpp_dir, &quot;/Poisson_just_fixed_expression&quot;)))
num_cores &lt;- 4</code></pre>
<pre class="r"><code>x &lt;- seq(0,4, length.out = 100)
a &lt;- 2*pi
sd_exact &lt;- 1
sd_quasi &lt;- 0.2
sd_np = 1
intercept_sd &lt;- 0.2
boundary_sd &lt;- 0.5
k &lt;- 50
m &lt;- 1
beta_prec &lt;- 1e-10
psd_vec &lt;- seq(0, 1, length.out = 11)
datasets &lt;- list()</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>We simulate the following data sets:</p>
<p>The first <span class="math inline">\(N_1\)</span> datasets are from
quasi-periodic sGP model:</p>
<pre class="r"><code>N1 &lt;- 100
sd_fun_draw_prop &lt;- c(0.2, 0.2, 0.15, 0.15, 0.15, 0.15)
sd_fun_draw_vec &lt;- rep(c(0, 0.1, 0.2, 0.3, 0.4, 0.5), times = c(0.2, 0.2, 0.15, 0.15, 0.15, 0.15)*N1)
for(i in 1:N1){
  sd_fun_draw &lt;- sd_fun_draw_vec[i]
  datasets[[i]] &lt;- simulate_data(type = &quot;v1&quot;, x = x, a = a, sd_fun = sd_fun_draw, boundary_sd = 1)
}
# plot 4 dataset
par(mfrow = c(2,2))
for(i in 1:4){
  index &lt;- sample(1:N1, 1)
  plot(datasets[[index]]$x, datasets[[index]]$y, type = &quot;p&quot;, col = &quot;blue&quot;, xlab = &quot;x&quot;, ylab = &quot;y&quot;)
  lines(datasets[[index]]$x, exp(datasets[[index]]$f), col = &quot;red&quot;, lwd = 2)
}</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-3-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>par(mfrow = c(1,1))</code></pre>
<p>The second <span class="math inline">\(N_2\)</span> datasets are from
the Poisson model with no periodic effect:</p>
<pre class="r"><code>N2 &lt;- 200
sd_fun_draw_prop &lt;- c(0.2, 0.2, 0.15, 0.15, 0.15, 0.15)
sd_fun_draw_vec2 &lt;- rep(c(0, 0.1, 0.2, 0.3, 0.4, 0.5), times = c(0.2, 0.2, 0.15, 0.15, 0.15, 0.15)*N2)
for(i in (N1+1):(N1+N2)){
  sd_fun_draw &lt;- sd_fun_draw_vec2[i-N1]
  datasets[[i]] &lt;- simulate_data(type = &quot;iwp&quot;, x = x, sd_fun = sd_fun_draw, intercept_sd = intercept_sd, boundary_sd = boundary_sd)
}
# plot 4 dataset
par(mfrow = c(2,2))
for(i in 1:4){
  index &lt;- sample((N1+1):(N1+N2), 1)
  plot(datasets[[index]]$x, datasets[[index]]$y, type = &quot;p&quot;, col = &quot;blue&quot;, xlab = &quot;x&quot;, ylab = &quot;y&quot;)
  lines(datasets[[index]]$x, exp(datasets[[index]]$f), col = &quot;red&quot;, lwd = 2)
}</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-4-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>par(mfrow = c(1,1))</code></pre>
<p>Compute the L matrix</p>
<pre class="r"><code>## set up a progress bar
L_vecs &lt;- vector(&quot;list&quot;, length(datasets))
pb &lt;- txtProgressBar(min = 0, max = length(datasets), style = 3)
for (i in 1:length(datasets)) {
  setTxtProgressBar(pb, i)
  data &lt;- datasets[[i]]
  like_vec1 &lt;- compute_log_likelihood_sBspline_seq(dataset = data, 
                                    period = 1, 
                                    num_knots = k, 
                                    psd_sgp_vector = psd_vec,
                                    pred_step = 1,
                                    betaprec = beta_prec, 
                                    log_lib_size = NULL, 
                                    m = 1)
  like_vec2 &lt;- compute_log_likelihood_ospline_seq(dataset = data, 
                                    p = 2, 
                                    num_knots = k, 
                                    psd_iwp_vector = psd_vec,
                                    pred_step = 1,
                                    betaprec = beta_prec, 
                                    log_lib_size = NULL)
  
  L_vecs[[i]] &lt;- c(like_vec1, like_vec2)
}
L_matrix &lt;- do.call(rbind, L_vecs)
save(L_matrix, file = paste0(result_dir, &quot;/L_matrix.rda&quot;))</code></pre>
<pre class="r"><code>load(paste0(result_dir, &quot;/L_matrix.rda&quot;))

fit.sqp &lt;- mixsqp(L = L_matrix, log = TRUE)</code></pre>
<pre><code>Running mix-SQP algorithm 0.3-54 on 300 x 22 matrix
convergence tol. (SQP):     1.0e-08
conv. tol. (active-set):    1.0e-10
zero threshold (solution):  1.0e-08
zero thresh. (search dir.): 1.0e-14
l.s. sufficient decrease:   1.0e-02
step size reduction factor: 7.5e-01
minimum step size:          1.0e-08
max. iter (SQP):            1000
max. iter (active-set):     20
number of EM iterations:    10
Computing SVD of 300 x 22 matrix.
Matrix is not low-rank; falling back to full matrix.
iter        objective max(rdual) nnz stepsize max.diff nqp nls
   1 +1.834688689e+00  -- EM --   22 1.00e+00 7.63e-02  --  --
   2 +1.809032472e+00  -- EM --   22 1.00e+00 1.86e-02  --  --
   3 +1.802387630e+00  -- EM --   22 1.00e+00 6.08e-03  --  --
   4 +1.799243370e+00  -- EM --   22 1.00e+00 2.96e-03  --  --
   5 +1.797338838e+00  -- EM --   22 1.00e+00 2.68e-03  --  --
   6 +1.796030452e+00  -- EM --   22 1.00e+00 2.41e-03  --  --
   7 +1.795059818e+00  -- EM --   22 1.00e+00 2.18e-03  --  --
   8 +1.794302358e+00  -- EM --   21 1.00e+00 1.98e-03  --  --
   9 +1.793689985e+00  -- EM --   21 1.00e+00 1.81e-03  --  --
  10 +1.793181842e+00  -- EM --   21 1.00e+00 1.66e-03  --  --
   1 +1.792751642e+00 +2.904e-02  21  ------   ------   --  --
   2 +1.787775842e+00 +2.458e-03  11 1.00e+00 8.55e-02  12   1
   3 +1.787663961e+00 +1.064e-02  12 1.00e+00 3.10e-03   7   1
   4 +1.787659796e+00 +3.961e-04  12 1.00e+00 8.07e-04   2   1
   5 +1.787659790e+00 +3.834e-07  12 1.00e+00 3.32e-05   2   1
   6 +1.787659790e+00 -4.276e-08  12 1.00e+00 5.07e-08   2   1
Optimization took 0.00 seconds.
Convergence criteria met---optimal solution found.</code></pre>
<pre class="r"><code>numiter &lt;- nrow(fit.sqp$progress)
plot(1:numiter,fit.sqp$progress$objective,type = &quot;b&quot;,
     pch = 20,lwd = 2,xlab = &quot;SQP iteration&quot;,
     ylab = &quot;objective&quot;,xaxp = c(1,numiter,numiter - 1))</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-6-1">
Past versions of unnamed-chunk-6-1.png
</button>
</p>
<div id="fig-unnamed-chunk-6-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-6-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code># create prior weight 
prior_weight &lt;- data.frame(name = paste0(rep(c(&quot;sgp_&quot;, &quot;iwp_&quot;), each = length(psd_vec)), psd_vec), prior_weight = fit.sqp$x)
prior_weight</code></pre>
<pre><code>      name prior_weight
1    sgp_0   0.07012691
2  sgp_0.1   0.05401808
3  sgp_0.2   0.06942097
4  sgp_0.3   0.06752890
5  sgp_0.4   0.00000000
6  sgp_0.5   0.07114683
7  sgp_0.6   0.00394424
8  sgp_0.7   0.00000000
9  sgp_0.8   0.00000000
10 sgp_0.9   0.00000000
11   sgp_1   0.00000000
12   iwp_0   0.11284080
13 iwp_0.1   0.18341235
14 iwp_0.2   0.06457638
15 iwp_0.3   0.13420174
16 iwp_0.4   0.00000000
17 iwp_0.5   0.14986793
18 iwp_0.6   0.01891489
19 iwp_0.7   0.00000000
20 iwp_0.8   0.00000000
21 iwp_0.9   0.00000000
22   iwp_1   0.00000000</code></pre>
<p>Compute the posterior:</p>
<pre class="r"><code># Compute the posterior using the likelihood and prior
posterior_matrix &lt;- matrix(0, nrow = nrow(L_matrix), ncol = ncol(L_matrix))
for(i in 1:nrow(L_matrix)){
  posterior_matrix[i,] &lt;- exp(L_matrix[i,] - max(L_matrix[i,]) + fit.sqp$x)
  posterior_matrix[i,] &lt;- posterior_matrix[i,]/sum(posterior_matrix[i,])
}</code></pre>
<pre class="r"><code>colnames(posterior_matrix) &lt;- paste0(rep(c(&quot;sgp_&quot;, &quot;iwp_&quot;), each = length(psd_vec)), psd_vec)</code></pre>
<p>The posterior probability of being in class sgp (summing all sgp
columns):</p>
<pre class="r"><code>posterior_prob &lt;- rowSums(posterior_matrix[, grepl(&quot;sgp_&quot;, colnames(posterior_matrix))])
head(posterior_prob)</code></pre>
<pre><code>[1] 1.0000000 0.3749057 1.0000000 1.0000000 1.0000000 1.0000000</code></pre>
<pre class="r"><code>plot(posterior_prob, type = &quot;p&quot;, col = &quot;blue&quot;, xlab = &quot;dataset&quot;, ylab = &quot;posterior probability&quot;)</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-10-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>posterior_weights_df &lt;- as.data.frame(posterior_matrix)
posterior_weights_df$id &lt;- 1:nrow(posterior_weights_df)
melted_data &lt;- melt(posterior_weights_df, id.vars = &quot;id&quot;)

ggplot(melted_data, aes(x = id, y = value, fill = variable)) +
  geom_bar(stat = &quot;identity&quot;, width = 0.7) +  # Adjust bar width if necessary
  ggtitle(&quot;Structure Plot of Posterior Weights&quot;) +
  coord_flip() +  # Flips the coordinates to make &#39;Gene&#39; on the y-axis
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),  
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA) 
  )</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-11-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The local FDR for testing periodicity:</p>
<pre class="r"><code>lfdr &lt;- 1 - posterior_prob
lfdr &lt;- ifelse(lfdr &lt; 0, 0, lfdr) # set to 0 if it is negative due to numerical error
lfdr &lt;- ifelse(lfdr &gt; 1, 1, lfdr) # set to 1 if it is greater than 1 due to numerical error

fdr_df &lt;- data.frame(ID = 1:length(lfdr), fdr = lfdr, type = rep(c(&quot;P&quot;, &quot;NP&quot;), times = c(N1,N2)))
fdr_df &lt;- fdr_df[order(fdr_df$fdr), ] # ordering it
fdr_df$cumulative_fdr &lt;- cumsum(fdr_df$fdr)/seq_along(fdr_df$fdr)
fdr_df$rank &lt;- 1:length(lfdr)</code></pre>
<p>Plot the FDR curve:</p>
<pre class="r"><code>ggplot(fdr_df, aes(x = 1:length(lfdr), y = cumulative_fdr, col = type)) +
  geom_point() +
  geom_hline(yintercept = 0.05, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Ordered Units&quot;, y = &quot;Cumulative FDR&quot;, col = &quot;Type&quot;) +
  theme_minimal() +
  ggtitle(&quot;FASH:Cumulative FDR Plot&quot;) +
  scale_color_manual(values = c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;))</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-13-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot the curve of nominal false discovery rate (threshold) against
the actual false discovery rate:</p>
<pre class="r"><code># Calculate true FDR for FASH
threshold_vec &lt;- seq(0, 2/3, by = 0.001)
fdr_vec &lt;- numeric(length(threshold_vec))

for (i in 1:length(threshold_vec)) {
  num_discoveries &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i] &amp; fdr_df$type == &quot;NP&quot;)
  fdr_vec[i] &lt;- num_false_discoveries / num_discoveries
}

# Create a data frame for plotting
fdr_df_fash_for_plotting &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec)

# Plot the nominal FDR vs true FDR for FASH
ggplot(fdr_df_fash_for_plotting, aes(x = threshold, y = true_fdr)) +
  geom_line() +
  geom_point(size = 0.1) +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;(Nominal) False Discovery Rate&quot;, y = &quot;(Actual) False Discovery Rate&quot;) +
  theme_minimal() +
  geom_hline(yintercept = N2/(N1 + N2), linetype = &quot;dashed&quot;, color = &quot;red&quot;) +
  coord_cartesian(xlim = c(0, 2/3), ylim = c(0, 2/3)) +
  ggtitle(&quot;Nominal FDR vs Actual FDR Curve for FASH&quot;)</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-14-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>threshold_vec &lt;- seq(0, 1, by = 0.01)
num_discoveries_vec &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec[i] &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries_vec[i] &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i] &amp; fdr_df$type == &quot;NP&quot;)
}
num_discoveries_df &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec, num_false_discoveries = num_false_discoveries_vec)
ggplot(num_discoveries_df, aes(x = (num_discoveries), y = (num_false_discoveries))) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Number of Discoveries&quot;, y = &quot;Number of False Discoveries&quot;) +
  theme_minimal() +
  ggtitle(&quot;FASH: Number of False Discoveries vs Number of Discoveries&quot;) +
  coord_cartesian(xlim = c(0, (N1 + N2)), ylim = c(0, (N2)))</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-1">
Past versions of unnamed-chunk-15-1.png
</button>
</p>
<div id="fig-unnamed-chunk-15-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-15-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>threshold_vec &lt;- seq(0, 1, by = 0.0001)
num_discoveries_vec_fash &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec_fash &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec_fash[i] &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries_vec_fash[i] &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i] &amp; fdr_df$type == &quot;NP&quot;)
}
num_discoveries_df_fash &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec_fash, num_false_discoveries = num_false_discoveries_vec_fash, method = &quot;FASH&quot;)

# Calculate the power for FASH
num_discoveries_df_fash$power &lt;- (num_discoveries_df_fash$num_discoveries-num_discoveries_df_fash$num_false_discoveries)/(N1)

# Plot the FDR vs Power
ggplot(num_discoveries_df_fash, aes(x = threshold, y = power, color = method)) +
  geom_line() +
  # geom_point() +
  labs(x = &quot;FDR&quot;, y = &quot;Power&quot;) +
  theme_minimal() +
  ggtitle(&quot;FDR vs Power for FASH&quot;) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1))</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-16-1">
Past versions of unnamed-chunk-16-1.png
</button>
</p>
<div id="fig-unnamed-chunk-16-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-16-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Compare with the posterior using the oracle Bayes:</p>
<p>Compute the posterior:</p>
<pre class="r"><code># Compute the posterior using the likelihood and prior
posterior_matrix_ob &lt;- matrix(0, nrow = nrow(L_matrix), ncol = ncol(L_matrix))
true_prior &lt;- c((N1/(N1+N2))*sd_fun_draw_prop, rep(0, (length(psd_vec) - length(sd_fun_draw_prop))) , (N2/(N1+N2))*sd_fun_draw_prop, rep(0, (length(psd_vec) - length(sd_fun_draw_prop))))
for(i in 1:nrow(L_matrix)){
  posterior_matrix_ob[i,] &lt;- exp(L_matrix[i,] - max(L_matrix[i,]) + log(true_prior))
  posterior_matrix_ob[i,] &lt;- posterior_matrix_ob[i,]/sum(posterior_matrix_ob[i,])
}</code></pre>
<pre class="r"><code>colnames(posterior_matrix_ob) &lt;- paste0(rep(c(&quot;sgp_&quot;, &quot;iwp_&quot;), each = length(psd_vec)), psd_vec)

posterior_prob_ob &lt;- rowSums(posterior_matrix_ob[, grepl(&quot;sgp_&quot;, colnames(posterior_matrix_ob))])
head(posterior_prob_ob)</code></pre>
<pre><code>[1] 1.0000000 0.2479439 1.0000000 1.0000000 1.0000000 1.0000000</code></pre>
<pre class="r"><code>plot(posterior_prob_ob, type = &quot;p&quot;, col = &quot;blue&quot;, xlab = &quot;dataset&quot;, ylab = &quot;posterior probability&quot;)</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-19-1">
Past versions of unnamed-chunk-19-1.png
</button>
</p>
<div id="fig-unnamed-chunk-19-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-19-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>posterior_weights_df_ob &lt;- as.data.frame(posterior_matrix_ob)

posterior_weights_df_ob$id &lt;- 1:nrow(posterior_weights_df_ob)
melted_data_ob &lt;- melt(posterior_weights_df_ob, id.vars = &quot;id&quot;)

ggplot(melted_data_ob, aes(x = id, y = value, fill = variable)) +
  geom_bar(stat = &quot;identity&quot;, width = 0.7) +  # Adjust bar width if necessary
  ggtitle(&quot;Structure Plot of Posterior Weights&quot;) +
  coord_flip() +  # Flips the coordinates to make &#39;Gene&#39; on the y-axis
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),  
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA) 
  )</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-19-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-19-2">
Past versions of unnamed-chunk-19-2.png
</button>
</p>
<div id="fig-unnamed-chunk-19-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-19-2.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The local FDR for testing periodicity:</p>
<pre class="r"><code>lfdr_ob &lt;- 1 - posterior_prob_ob
lfdr_ob &lt;- ifelse(lfdr_ob &lt; 0, 0, lfdr_ob) # set to 0 if it is negative due to numerical error
lfdr_ob &lt;- ifelse(lfdr_ob &gt; 1, 1, lfdr_ob) # set to 1 if it is greater than 1 due to numerical error


fdr_df_ob &lt;- data.frame(ID = 1:length(lfdr_ob), fdr = lfdr_ob, type = rep(c(&quot;P&quot;, &quot;NP&quot;), times = c(N1,N2)))
fdr_df_ob &lt;- fdr_df_ob[order(fdr_df_ob$fdr), ] # ordering it
fdr_df_ob$cumulative_fdr &lt;- cumsum(fdr_df_ob$fdr)/seq_along(fdr_df_ob$fdr)
fdr_df_ob$rank &lt;- 1:length(lfdr_ob)

ggplot(fdr_df_ob, aes(x = 1:length(lfdr_ob), y = cumulative_fdr, col = type)) +
  geom_point() +
  geom_hline(yintercept = 0.05, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Ordered Units&quot;, y = &quot;Cumulative FDR&quot;, col = &quot;Type&quot;) +
  theme_minimal() +
  ggtitle(&quot;Oracle: Cumulative FDR Plot&quot;) +
  scale_color_manual(values = c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;))</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-20-1">
Past versions of unnamed-chunk-20-1.png
</button>
</p>
<div id="fig-unnamed-chunk-20-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-20-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Check calibration of oracle Bayes: the curve of nominal false
discovery rate (threshold) against the actual false discovery rate:</p>
<pre class="r"><code># Calculate true FDR for Oracle
threshold_vec &lt;- seq(0, 2/3, by = 0.001)
fdr_vec_ob &lt;- numeric(length(threshold_vec))

for (i in 1:length(threshold_vec)) {
  num_discoveries &lt;- sum(fdr_df_ob$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries &lt;- sum(fdr_df_ob$cumulative_fdr &lt;= threshold_vec[i] &amp; fdr_df_ob$type == &quot;NP&quot;)
  fdr_vec_ob[i] &lt;- num_false_discoveries / num_discoveries
}

# Create a data frame for plotting

fdr_df_ob_for_plotting &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_ob)

# Plot the nominal FDR vs true FDR for Oracle

ggplot(fdr_df_ob_for_plotting, aes(x = threshold, y = true_fdr)) +
  geom_line() +
  geom_point(size = 0.1) +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;(Nominal) False Discovery Rate&quot;, y = &quot;(Actual) False Discovery Rate&quot;) +
  theme_minimal() +
  geom_hline(yintercept = N2/(N1 + N2), linetype = &quot;dashed&quot;, color = &quot;red&quot;) +
  coord_cartesian(xlim = c(0, 2/3), ylim = c(0, 2/3)) +
  ggtitle(&quot;Nominal FDR vs Actual FDR Curve for Oracle&quot;)</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-21-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>threshold_vec &lt;- seq(0, 1, by = 0.01)
num_discoveries_vec_ob &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec_ob &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec_ob[i] &lt;- sum(fdr_df_ob$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries_vec_ob[i] &lt;- sum(fdr_df_ob$cumulative_fdr &lt;= threshold_vec[i] &amp; fdr_df_ob$type == &quot;NP&quot;)
}
num_discoveries_df_ob &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec_ob, num_false_discoveries = num_false_discoveries_vec_ob)
ggplot(num_discoveries_df_ob, aes(x = (num_discoveries), y = (num_false_discoveries))) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Number of Discoveries&quot;, y = &quot;Number of False Discoveries&quot;) +
  theme_minimal() +
  ggtitle(&quot;Oracle: Number of False Discoveries vs Number of Discoveries&quot;) +
  coord_cartesian(xlim = c(0, (N1 + N2)), ylim = c(0, (N2)))</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-22-1.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>threshold_vec &lt;- seq(0, 1, by = 0.0001)
num_discoveries_vec_ob_fash &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec_ob_fash &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec_ob_fash[i] &lt;- sum(fdr_df_ob$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries_vec_ob_fash[i] &lt;- sum(fdr_df_ob$cumulative_fdr &lt;= threshold_vec[i] &amp; fdr_df_ob$type == &quot;NP&quot;)
}
num_discoveries_df_ob_fash &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec_ob_fash, num_false_discoveries = num_false_discoveries_vec_ob_fash, method = &quot;Oracle&quot;)

# Calculate the power for Oracle
num_discoveries_df_ob_fash$power &lt;- (num_discoveries_df_ob_fash$num_discoveries-num_discoveries_df_ob_fash$num_false_discoveries)/(N1)

# Plot the FDR vs Power

ggplot(num_discoveries_df_ob_fash, aes(x = threshold, y = power, color = method)) +
  geom_line() +
  # geom_point() +
  labs(x = &quot;FDR&quot;, y = &quot;Power&quot;) +
  theme_minimal() +
  ggtitle(&quot;FDR vs Power for Oracle&quot;) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1))</code></pre>
<p><img src="figure/circadian_sim_easy.Rmd/unnamed-chunk-22-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-2">
Past versions of unnamed-chunk-22-2.png
</button>
</p>
<div id="fig-unnamed-chunk-22-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d634fe361609d2dc92204afc5397e787ef9ad2e3/docs/figure/circadian_sim_easy.Rmd/unnamed-chunk-22-2.png" target="_blank">d634fe3</a>
</td>
<td>
AgueroZZ
</td>
<td>
2024-09-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.0   dplyr_1.1.3    
 [5] purrr_1.0.2     readr_2.1.4     tidyr_1.3.0     tibble_3.2.1   
 [9] tidyverse_2.0.0 mixsqp_0.3-54   reshape2_1.4.4  ggplot2_3.5.1  
[13] Matrix_1.6-3    TMB_1.9.7       BayesGP_0.1.1  

loaded via a namespace (and not attached):
 [1] gtable_0.3.4         xfun_0.40            bslib_0.5.1         
 [4] ks_1.14.1            lattice_0.21-8       tzdb_0.4.0          
 [7] bitops_1.0-7         vctrs_0.6.4          tools_4.3.1         
[10] generics_0.1.3       fansi_1.0.5          cluster_2.1.4       
[13] pkgconfig_2.0.3      fds_1.8              KernSmooth_2.23-21  
[16] lifecycle_1.0.3      farver_2.1.1         compiler_4.3.1      
[19] git2r_0.33.0         munsell_0.5.0        httpuv_1.6.11       
[22] htmltools_0.5.6.1    rainbow_3.7          sass_0.4.7          
[25] RCurl_1.98-1.12      yaml_2.3.7           pracma_2.4.2        
[28] later_1.3.1          pillar_1.9.0         jquerylib_0.1.4     
[31] whisker_0.4.1        MASS_7.3-60          cachem_1.0.8        
[34] mclust_6.0.0         tidyselect_1.2.0     digest_0.6.33       
[37] mvtnorm_1.2-3        stringi_1.7.12       LaplacesDemon_16.1.6
[40] labeling_0.4.3       pcaPP_2.0-3          rprojroot_2.0.3     
[43] fastmap_1.1.1        grid_4.3.1           colorspace_2.1-0    
[46] cli_3.6.1            magrittr_2.0.3       utf8_1.2.3          
[49] withr_2.5.1          scales_1.3.0         promises_1.2.1      
[52] timechange_0.2.0     rmarkdown_2.25       deSolve_1.38        
[55] workflowr_1.7.1      hms_1.1.3            evaluate_0.22       
[58] knitr_1.44           irlba_2.3.5.1        rlang_1.1.1         
[61] Rcpp_1.0.11          hdrcde_3.4           glue_1.6.2          
[64] fda_6.1.4            rstudioapi_0.15.0    jsonlite_1.8.7      
[67] R6_2.5.1             plyr_1.8.9           fs_1.6.3            </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
