<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Example: Expression Analysis</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">resultsummary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="simulation_pollution.html">Simulation 1</a>
</li>
<li>
  <a href="eQTL_sim.html">Simulation 2</a>
</li>
<li>
  <a href="covid_example.html">COVID-19 Example</a>
</li>
<li>
  <a href="expression.html">Gene Expression Example</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/AgueroZZ/resultsummary">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Example: Expression Analysis</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-10-08
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>FASHresultsummary/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240507code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240507)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240507code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240507)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytree2946863baa8d6b68a930084aa44e1565911e7b5atargetblank2946863a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/2946863baa8d6b68a930084aa44e1565911e7b5a" target="_blank">2946863</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytree2946863baa8d6b68a930084aa44e1565911e7b5atargetblank2946863a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/2946863baa8d6b68a930084aa44e1565911e7b5a" target="_blank">2946863</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    code/.DS_Store
    Ignored:    code/cpp/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/expression_data/.DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/example/.DS_Store
    Ignored:    output/example/figure/
    Ignored:    output/simulation_pollution/.DS_Store
    Ignored:    output/simulation_pollution/figure/

Untracked files:
    Untracked:  analysis/circadian_expression (experimment).Rmd
    Untracked:  output/expression/gene_set_table_strober

Unstaged changes:
    Modified:   analysis/circadian_expression.Rmd
    Modified:   code/function/functions_fitting_Poisson_circadian_expression.R
    Modified:   output/circadian_expression/L_matrix.rda
    Modified:   output/circadian_sim_easy/L_matrix.rda
    Modified:   output/expression/L_matrix.rda
    Modified:   thought_process

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/expression.Rmd</code>) and HTML
(<code>docs/expression.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/2946863baa8d6b68a930084aa44e1565911e7b5a/analysis/expression.Rmd" target="_blank">2946863</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-08
</td>
<td>
wflow_publish("analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/6fc89dfc0f1c89476b37b961daf199bc72ec1ff6/docs/expression.html" target="_blank">6fc89df</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/f8d32d7614879dd849c02a80bcbdb6686636928d/analysis/expression.Rmd" target="_blank">f8d32d7</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-07
</td>
<td>
wflow_publish("analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/719af82763637ca6dd4dd8b318d656ecea79da60/docs/expression.html" target="_blank">719af82</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d78bfa8bb385310fe969df41842b6fc02ff257cd/analysis/expression.Rmd" target="_blank">d78bfa8</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-02
</td>
<td>
wflow_publish("analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/ec579c2dfa555a1289dbd88d16fa8a6475ce0632/docs/expression.html" target="_blank">ec579c2</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/0414ce6e5ba33711324308e3be187f47e267cb3d/analysis/expression.Rmd" target="_blank">0414ce6</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-01
</td>
<td>
workflowr::wflow_publish("analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/ad988fcf8567dbcc95a1673c376253076d7a23a5/docs/expression.html" target="_blank">ad988fc</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/e57fc389bf9a834c1770a720a2cc554ee769424e/analysis/expression.Rmd" target="_blank">e57fc38</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
<td>
wflow_publish("analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a22557d0109871a2740522e971cb8f72c87ac72b/analysis/expression.Rmd" target="_blank">a22557d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
<td>
update the website
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/a22557d0109871a2740522e971cb8f72c87ac72b/docs/expression.html" target="_blank">a22557d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
<td>
update the website
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/b729c3febd0f9dc9309871c6d9ce8544ecff7d5e/docs/expression.html" target="_blank">b729c3f</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-06-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/62566ff29f2e98ea1b83dcba8b8ad3a569e58d41/analysis/expression.Rmd" target="_blank">62566ff</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-06-10
</td>
<td>
wflow_git_commit(c("analysis/expression.Rmd"))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/fb94e2d4da03316e6e8a5dc0c16e25808d539482/docs/expression.html" target="_blank">fb94e2d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-31
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d6a5a69e5ec3b988d534ca72b051ef520e198899/analysis/expression.Rmd" target="_blank">d6a5a69</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-31
</td>
<td>
wflow_publish("analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/bf7911aace86fe5070c0698f4e741784617c7b50/docs/expression.html" target="_blank">bf7911a</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-31
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d7ad1da14c525fb5589f5288657046c989edd6d9/analysis/expression.Rmd" target="_blank">d7ad1da</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-31
</td>
<td>
wflow_publish("analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/6d68a5df7e7473f23551b0c3f25f37d001701372/docs/expression.html" target="_blank">6d68a5d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/da46108aba2e0e5b1756365bbb0fc54cf2f75b11/analysis/expression.Rmd" target="_blank">da46108</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
<td>
wflow_publish(files = "analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/00d489f81ae68aabc7c4692348649dcdde6dcaae/docs/expression.html" target="_blank">00d489f</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-27
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ade221629635a11d5d73a18ffd7c78db00571d31/analysis/expression.Rmd" target="_blank">ade2216</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-27
</td>
<td>
wflow_publish(files = "analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/cfe24a9e633b867d989d93e3874b3d7a6c9d4f94/docs/expression.html" target="_blank">cfe24a9</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-27
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/16264cf81cf5ebdf91b997f83c280d55d9222490/analysis/expression.Rmd" target="_blank">16264cf</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-27
</td>
<td>
wflow_publish(files = "analysis/expression.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/425cb9dd1489c78874b8e8e07396d5cf339664b4/docs/expression.html" target="_blank">425cb9d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-27
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ab342cbde768b4282e6db6a241018b33444b998a/analysis/expression.Rmd" target="_blank">ab342cb</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-27
</td>
<td>
wflow_git_commit(c("analysis/simulation_eQTL_identification.Rmd",
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/3b37c543448424223e37bef6f636d8c9dce217ec/docs/expression.html" target="_blank">3b37c54</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
update html
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/08cc7a7aebe5d38aa45f16a4164627dc04555198/analysis/expression.Rmd" target="_blank">08cc7a7</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
wflow_git_commit(c("analysis/expression.Rmd",
"analysis/covid_example.Rmd"))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/expression.html" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
update the expression example
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/f571ee01ac5a385ecb0d9a7678a133062d970451/analysis/expression.Rmd" target="_blank">f571ee0</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
wflow_git_commit(c("analysis/covid_example.Rmd",
"code/function/functions_fitting_Poisson_covid.R",
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="setup" class="section level3">
<h3><strong>Setup</strong></h3>
<p>We consider the (raw) gene expression data measured over 16 days as
studied in <a
href="https://www.science.org/doi/10.1126/science.aaw0040">Strober et
al, 2019</a>. For simplicity of the presentation, we consider only one
cell-line in this example. The dataset contains the raw expression
counts of 16319 genes.</p>
<p>To model the true (log) expression level <span
class="math inline">\(f_i(t_j)\)</span> of the <span
class="math inline">\(i\)</span>th gene at day <span
class="math inline">\(t_j\)</span>, we assume that the expression count
<span class="math inline">\(y_{i}(t_j)\)</span> is Poisson distributed
with mean <span class="math inline">\(o_j\exp
\left(f_i\left(t_j\right)\right)\)</span>, where <span
class="math inline">\(o_j\)</span> is the library size at day <span
class="math inline">\(t_j\)</span>.</p>
</div>
<div id="model" class="section level3">
<h3><strong>Model</strong></h3>
<p>Assume most of the gene expression trajectories are expected to be
linear, we therefore consider the following finite mixture prior for
each gene expression level <span
class="math inline">\(f_i(t_j)\)</span>: <span
class="math display">\[f_i(t) = \sum_{k=0}^K \pi_k
\text{GP}_k(t).\]</span> For each <span
class="math inline">\(k\)</span>, the mixture component has the prior:
<span class="math display">\[Lf(t) = \sigma_k W(t), \]</span> with <span
class="math inline">\(W(t)\)</span> being a Gaussian white noise process
and <span class="math inline">\(\sigma_k\)</span> being the smoothness.
Choosing <span class="math inline">\(L = \frac{\partial^2}{\partial
t^2}\)</span>, the base model that each <span
class="math inline">\(f_i\)</span> being shrunk to is therefore <span
class="math inline">\(\text{Null}\{L\} =
\text{span}\{1,t\}\)</span>.</p>
</div>
<div id="processing-the-expression-data" class="section level3">
<h3><strong>Processing the expression data</strong></h3>
<pre class="r"><code>all_data_names &lt;- list.files(data_dir)
#### let&#39;s consider one cell-line for simplicity
load(paste0(data_dir, &quot;/&quot;, all_data_names[1]))
str(expression_data_one_cell_line) </code></pre>
<pre><code>tibble [261,104 × 5] (S3: tbl_df/tbl/data.frame)
 $ Gene_id           : chr [1:261104] &quot;ENSG00000000419&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000419&quot; ...
 $ Cell_line         : chr [1:261104] &quot;18489&quot; &quot;18489&quot; &quot;18489&quot; &quot;18489&quot; ...
 $ Day               : int [1:261104] 0 1 2 3 4 5 6 7 8 9 ...
 $ value             : int [1:261104] 825 2432 1044 874 766 632 839 685 640 636 ...
 $ external_gene_name: chr [1:261104] &quot;DPM1&quot; &quot;DPM1&quot; &quot;DPM1&quot; &quot;DPM1&quot; ...</code></pre>
<p>Let’s create one dataset for each gene:</p>
<pre class="r"><code>num_knots &lt;- 16
p &lt;- 2
datasets &lt;- list()
all_gene &lt;- unique(expression_data_one_cell_line$Gene_id)
for (gene in all_gene) {
  datasets[[gene]] &lt;- expression_data_one_cell_line %&gt;% filter(Gene_id == gene)
  datasets[[gene]]$x &lt;- as.numeric(datasets[[gene]]$Day)
  datasets[[gene]]$y &lt;- datasets[[gene]]$value
}
save(datasets, file = paste0(result_dir, &quot;/datasets.rda&quot;))</code></pre>
<p>Compute the library sizes:</p>
<pre class="r"><code>size_vec &lt;- numeric(length = nrow(datasets[[1]]))
for (i in 1:length(size_vec)) {
  all_counts &lt;- unlist(lapply(datasets, function(x) {x$y[x$Day == (i-1)]}))
  size_vec[i] &lt;- sum(all_counts)
}
log_size_vec &lt;- log(size_vec)</code></pre>
</div>
<div id="applying-empirical-bayes" class="section level3">
<h3><strong>Applying empirical Bayes</strong></h3>
<pre class="r"><code>psd_vec &lt;- sort(unique(c(0, seq(0,3, length.out = 31))))

## set up a progress bar
L_vecs &lt;- vector(&quot;list&quot;, length(datasets))
pb &lt;- txtProgressBar(min = 0, max = length(datasets), style = 3)
for (i in 1:length(datasets)) {
  setTxtProgressBar(pb, i)
  dataset &lt;- datasets[[i]]
  L_vecs[[i]] &lt;- compute_log_likelihood_ospline_seq2(
    dataset = dataset,
    p = p,
    num_knots = num_knots,
    psd_iwp_vector = psd_vec,
    pred_step = 1,
    betaprec = betaprec,
    log_lib_size = log_size_vec
  )
}
L_matrix &lt;- do.call(rbind, L_vecs)
save(L_matrix, file = paste0(result_dir, &quot;/L_matrix.rda&quot;))</code></pre>
<pre class="r"><code>fit.sqp &lt;- mixsqp(L = L_matrix, log = TRUE)</code></pre>
<pre><code>Running mix-SQP algorithm 0.3-54 on 16319 x 31 matrix
convergence tol. (SQP):     1.0e-08
conv. tol. (active-set):    1.0e-10
zero threshold (solution):  1.0e-08
zero thresh. (search dir.): 1.0e-14
l.s. sufficient decrease:   1.0e-02
step size reduction factor: 7.5e-01
minimum step size:          1.0e-08
max. iter (SQP):            1000
max. iter (active-set):     20
number of EM iterations:    10
Computing SVD of 16319 x 31 matrix.
Matrix is not low-rank; falling back to full matrix.
iter        objective max(rdual) nnz stepsize max.diff nqp nls
   1 +1.634441920e+00  -- EM --   31 1.00e+00 1.11e-01  --  --
   2 +1.618829189e+00  -- EM --   31 1.00e+00 2.75e-02  --  --
   3 +1.615448617e+00  -- EM --   31 1.00e+00 9.87e-03  --  --
   4 +1.613879324e+00  -- EM --   31 1.00e+00 5.46e-03  --  --
   5 +1.612942812e+00  -- EM --   31 1.00e+00 3.48e-03  --  --
   6 +1.612296643e+00  -- EM --   31 1.00e+00 2.42e-03  --  --
   7 +1.611809877e+00  -- EM --   31 1.00e+00 2.10e-03  --  --
   8 +1.611423134e+00  -- EM --   31 1.00e+00 1.84e-03  --  --
   9 +1.611105393e+00  -- EM --   31 1.00e+00 1.61e-03  --  --
  10 +1.610838410e+00  -- EM --   31 1.00e+00 1.42e-03  --  --
   1 +1.610610420e+00 +3.320e-02  31  ------   ------   --  --
   2 +1.608587903e+00 +2.507e-03  13 1.00e+00 6.29e-02  20   1
   3 +1.608291005e+00 +5.191e-04  13 1.00e+00 1.04e-01  20   1
   4 +1.608290952e+00 +1.919e-07  14 1.00e+00 3.60e-03  12   1
   5 +1.608290952e+00 -4.218e-08  14 1.00e+00 2.31e-06  18   1
Optimization took 0.09 seconds.
Convergence criteria met---optimal solution found.</code></pre>
<pre class="r"><code>numiter &lt;- nrow(fit.sqp$progress)
plot(1:numiter,fit.sqp$progress$objective,type = &quot;b&quot;,
     pch = 20,lwd = 2,xlab = &quot;SQP iteration&quot;,
     ylab = &quot;objective&quot;,xaxp = c(1,numiter,numiter - 1))</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ec579c2dfa555a1289dbd88d16fa8a6475ce0632/docs/figure/expression.Rmd/unnamed-chunk-8-1.png" target="_blank">ec579c2</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ad988fcf8567dbcc95a1673c376253076d7a23a5/docs/figure/expression.Rmd/unnamed-chunk-8-1.png" target="_blank">ad988fc</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a22557d0109871a2740522e971cb8f72c87ac72b/docs/figure/expression.Rmd/unnamed-chunk-8-1.png" target="_blank">a22557d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/bf7911aace86fe5070c0698f4e741784617c7b50/docs/figure/expression.Rmd/unnamed-chunk-8-1.png" target="_blank">bf7911a</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-31
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>prior_weight &lt;- data.frame(p = rep(p, each = length(psd_vec)), psd_iwp = psd_vec, prior_weight = fit.sqp$x)

posterior_matrix &lt;- matrix(0, nrow = nrow(L_matrix), ncol = ncol(L_matrix))
for(i in 1:nrow(L_matrix)){
  posterior_matrix[i,] &lt;- exp(L_matrix[i,] - max(L_matrix[i,]) + log(fit.sqp$x))
  posterior_matrix[i,] &lt;- posterior_matrix[i,]/sum(posterior_matrix[i,])
}
colnames(posterior_matrix) &lt;- paste0(p,&quot;_&quot;,psd_vec)
posterior_weights_matrix &lt;- posterior_matrix</code></pre>
</div>
<div id="obtaining-posteriors" class="section level3">
<h3><strong>Obtaining posteriors</strong></h3>
<pre class="r"><code>num_datasets &lt;- length(datasets)
num_weights &lt;- sum(prior_weight$prior_weight != 0)
posterior_weights_matrix &lt;- matrix(nrow = num_datasets, ncol = num_weights)

# Loop through each dataset and perform fitting
fitted_datasets &lt;- list()
# create a progress bar
pb &lt;- txtProgressBar(min = 0, max = num_datasets, style = 3)
for (i in seq_along(datasets)) {
  setTxtProgressBar(pb, i)
  dataset &lt;- datasets[[i]]
  fit_result_final &lt;- fit_ospline_with_prior2(
    num_cores = 1,
    dataset = dataset,
    num_knots = num_knots,
    prior_weight = prior_weight,
    betaprec = betaprec,
    pred_step = 1,
    log_lib_size = log_size_vec
  )
  posterior_weights_matrix[i, ] &lt;- fit_result_final$posterior_weights[, &quot;posterior_weight&quot;]
  fitted_datasets[[i]] &lt;- aggregate_fit_with_prior(x = dataset$x, fit_results_with_prior = fit_result_final, original = TRUE)$summary_df
}
names(fitted_datasets) &lt;- all_gene
colnames(posterior_weights_matrix) &lt;- paste(as.character(fit_result_final$posterior_weights[, &quot;p&quot;]),
                                            as.character(fit_result_final$posterior_weights[, &quot;psd_iwp&quot;]), sep = &quot;_&quot;)
save(fitted_datasets, posterior_weights_matrix, file = paste0(result_dir, &quot;/fitted_pos_results.rda&quot;))</code></pre>
<pre class="r"><code>posterior_weights_df &lt;- as.data.frame(posterior_weights_matrix)
posterior_weights_df$id &lt;- all_gene
melted_data &lt;- melt(posterior_weights_df, id.vars = &quot;id&quot;)
melted_data$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data$variable)
melted_data$variable3 &lt;- (round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data$variable)), 3))</code></pre>
<p>Plot a sorted version of this, based on the mean PSD in each
gene:</p>
<pre class="r"><code>mean_psd &lt;- apply(posterior_weights_matrix, 1, function(x) {
  weighted_psd &lt;- as.numeric(sub(&quot;*._&quot;, &quot;&quot;, colnames(posterior_weights_matrix)))
  sum(x * weighted_psd)
})
sorted_index &lt;- order(mean_psd)
sorted_posterior_weights_matrix &lt;- posterior_weights_matrix[sorted_index, ]
sorted_posterior_weights_df &lt;- as.data.frame(sorted_posterior_weights_matrix)
sorted_posterior_weights_df$id &lt;- all_gene[sorted_index]
melted_data_sorted &lt;- melt(sorted_posterior_weights_df, id.vars = &quot;id&quot;)
melted_data_sorted$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data_sorted$variable)
melted_data_sorted$variable3 &lt;- (round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data_sorted$variable)), 3))

ggplot(melted_data_sorted, aes(x = factor(id, levels = sorted_posterior_weights_df$id), y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +
  labs(x = &quot;Gene&quot;, y = &quot;Weight&quot;, fill = &quot;PSD&quot;) +
  ggtitle(&quot;Structure Plot of Posterior Weights&quot;) +
  coord_flip() +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),  
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA) 
  )</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ec579c2dfa555a1289dbd88d16fa8a6475ce0632/docs/figure/expression.Rmd/unnamed-chunk-11-1.png" target="_blank">ec579c2</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ad988fcf8567dbcc95a1673c376253076d7a23a5/docs/figure/expression.Rmd/unnamed-chunk-11-1.png" target="_blank">ad988fc</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a22557d0109871a2740522e971cb8f72c87ac72b/docs/figure/expression.Rmd/unnamed-chunk-11-1.png" target="_blank">a22557d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/bf7911aace86fe5070c0698f4e741784617c7b50/docs/figure/expression.Rmd/unnamed-chunk-11-1.png" target="_blank">bf7911a</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-11-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="clustering-based-on-posterior-weights" class="section level3">
<h3><strong>Clustering based on posterior weights</strong></h3>
<p>Let’s group them into <code>num_clusters</code> groups, based on the
quantiles of the mean PSD:</p>
<pre class="r"><code>mean_psd_ordered &lt;- mean_psd[sorted_index]
# num_clusters &lt;- 22
# quantiles &lt;- quantile(mean_psd_ordered, probs = seq(0, 1, 1/num_clusters))
quantiles &lt;- c(0, 0.1, 0.2, 0.35, 0.5, 1, 1.7, 2.5, 2.95, 3)

quantile_groups &lt;- cut(mean_psd_ordered, breaks = quantiles, 
                       include.lowest = TRUE, labels = FALSE)
table(quantile_groups)</code></pre>
<pre><code>quantile_groups
   1    2    3    4    5    6    7    8    9 
 446 3429 3168 2203 3879 1826  741  277  350 </code></pre>
<pre class="r"><code>num_clusters &lt;- length(unique(quantile_groups))

sorted_posterior_weights_df$cluster &lt;- quantile_groups
melted_data_quantiles &lt;- melt(sorted_posterior_weights_df, id.vars = c(&quot;id&quot;, &quot;cluster&quot;))
melted_data_quantiles$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data_quantiles$variable)
melted_data_quantiles$variable3 &lt;- round(as.numeric(sub(&quot;.*_&quot;, &quot;&quot;, melted_data_quantiles$variable)), 3)
melted_data_quantiles &lt;- melted_data_quantiles %&gt;% arrange(cluster)

# Ensure the &#39;id&#39; factor is ordered according to the sorted mean_psd
melted_data_quantiles$id &lt;- factor(melted_data_quantiles$id, levels = sorted_posterior_weights_df$id)

# Plotting with facets by cluster and proper sorting of &#39;id&#39;
ggplot(melted_data_quantiles, aes(x = factor(id, levels = sorted_posterior_weights_df$id), y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +
  labs(x = &quot;Gene&quot;, y = &quot;Weight&quot;, fill = &quot;PSD Level&quot;) +
  ggtitle(&quot;Structure Plot of Posterior Weights by Quantile Group&quot;) +
  coord_flip() +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  facet_wrap(~cluster, scales = &quot;free_y&quot;) +  # Faceting by cluster while keeping order
  theme_minimal() +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA),
    strip.background = element_rect(fill = &quot;lightblue&quot;, colour = &quot;deepskyblue&quot;, size = 1),
    strip.text = element_text(face = &quot;bold&quot;)
  )</code></pre>
<pre><code>Warning: The `size` argument of `element_rect()` is deprecated as of ggplot2 3.4.0.
ℹ Please use the `linewidth` argument instead.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-1">
Past versions of unnamed-chunk-12-1.png
</button>
</p>
<div id="fig-unnamed-chunk-12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/719af82763637ca6dd4dd8b318d656ecea79da60/docs/figure/expression.Rmd/unnamed-chunk-12-1.png" target="_blank">719af82</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-02
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ec579c2dfa555a1289dbd88d16fa8a6475ce0632/docs/figure/expression.Rmd/unnamed-chunk-12-1.png" target="_blank">ec579c2</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-10-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ad988fcf8567dbcc95a1673c376253076d7a23a5/docs/figure/expression.Rmd/unnamed-chunk-12-1.png" target="_blank">ad988fc</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a22557d0109871a2740522e971cb8f72c87ac72b/docs/figure/expression.Rmd/unnamed-chunk-12-1.png" target="_blank">a22557d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-09-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/bf7911aace86fe5070c0698f4e741784617c7b50/docs/figure/expression.Rmd/unnamed-chunk-12-1.png" target="_blank">bf7911a</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-12-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<!-- ### **Smoothing result for each cluster (Overall expression level $o_j\exp(f_i(t_j))$** {.tabset .tabset-pills} -->
<!-- ```{r, results='asis'} -->
<!-- for (i in 1:num_clusters) { -->
<!--   cat(paste0("#### For cluster ", i, ":\n")) -->
<!--   # Dynamically evaluate the plotting function for each cluster -->
<!--   plot_cluster_genes(genes_per_cluster[[i]], log_size_vec) -->
<!-- } -->
<!-- ``` -->
<!-- ### **Smoothing result for each cluster (Just the function $f_i$ after Log)** {.tabset .tabset-pills} -->
<!-- ```{r} -->
<!-- for (i in 1:num_clusters) { -->
<!--   # Header for each cluster -->
<!--   cat(paste0("#### For cluster ", i, ":\n")) -->
<!--   # Dynamically evaluate the plotting function for each cluster -->
<!--   plot_cluster_genes_log_just_f(genes_per_cluster[[i]], log_size_vec) -->
<!-- } -->
<!-- ``` -->
</div>
<div id="question-what-genes-are-enriched-in-each-of-the-cluster"
class="section level3">
<h3><strong>Question: what genes are enriched in each of the cluster?
</strong></h3>
<pre class="r"><code>library(clusterProfiler)
library(msigdbr)
library(org.Hs.eg.db)  # Assuming human genes
library(biomaRt)
library(clusterProfiler)
library(cowplot)

# Retrieve Hallmark gene sets for Homo sapiens
m_t2g &lt;- msigdbr(species = &quot;Homo sapiens&quot;, category = &quot;H&quot;) %&gt;% 
  dplyr::select(gs_name, entrez_gene)
mart &lt;- useMart(&quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)

enrich_cluster &lt;- function(cluster_id, q_val_cutoff = 0.05, pvalueCutoff = 0.05, melted_data_input) {
  genes_selected &lt;- melted_data_input %&gt;% filter(cluster == cluster_id) %&gt;% pull(id) %&gt;% as.character() %&gt;% unique()
  
  genes_converted &lt;- getBM(
  filters = &quot;ensembl_gene_id&quot;, 
  attributes = c(&quot;ensembl_gene_id&quot;, &quot;entrezgene_id&quot;), 
  values = genes_selected, 
  mart = mart
  )
  
  # Extract Entrez IDs from the converted data
  entrez_gene_list &lt;- genes_converted$entrezgene_id
  
  genes_converted_all &lt;- getBM(
    filters = &quot;ensembl_gene_id&quot;, 
    attributes = c(&quot;ensembl_gene_id&quot;, &quot;entrezgene_id&quot;), 
    values = melted_data_input$id %&gt;% as.character() %&gt;% unique(), 
    mart = mart
  )
  entrez_universe &lt;- as.character(genes_converted_all$entrezgene_id)
  entrez_universe &lt;- entrez_universe[!is.na(entrez_universe)]

  # Perform enrichment analysis using Hallmark gene sets
  enrich_res &lt;- enricher(pAdjustMethod = &quot;BH&quot;, 
                         entrez_gene_list, 
                         TERM2GENE = m_t2g, 
                         qvalueCutoff = q_val_cutoff, 
                         pvalueCutoff = pvalueCutoff, 
                         universe = entrez_universe)
  enrich_res
}</code></pre>
<pre class="r"><code>enriched_list &lt;- list()
for (i in 1:num_clusters) {
  # for each cluster, record the list of its enriched gene sets
  enrich_res &lt;- enrich_cluster(cluster_id = i, q_val_cutoff = 0.05, melted_data_input = melted_data_quantiles)  # Adjust parameters as needed
  enriched_list[[i]] &lt;- enrich_res$Description
  names(enriched_list)[i] &lt;- paste0(i)
}

# Automatically create a list of all unique enriched terms
all_enriched_terms &lt;- unique(unlist(enriched_list))

# Remove the &quot;HALLMARK_&quot; prefix from each term
all_enriched_terms_clean &lt;- gsub(&quot;HALLMARK_&quot;, &quot;&quot;, all_enriched_terms)

# Initialize a data frame with clusters as columns and terms as rows
enrichment_table &lt;- data.frame(matrix(nrow = length(all_enriched_terms_clean), ncol = length(enriched_list)))
rownames(enrichment_table) &lt;- all_enriched_terms_clean
colnames(enrichment_table) &lt;- names(enriched_list)

# Fill the table with &quot;yes&quot; or &quot;no&quot;
for (cluster in names(enriched_list)) {
  for (term in all_enriched_terms_clean) {
    # Check if the term is present in the enriched list of the cluster (after removing &quot;HALLMARK_&quot;)
    original_term &lt;- paste0(&quot;HALLMARK_&quot;, term)  # Add back the prefix for comparison
    enrichment_table[term, cluster] &lt;- ifelse(original_term %in% enriched_list[[cluster]], &quot;yes&quot;, &quot;no&quot;)
  }
}

# Display the resulting table
print(enrichment_table[,-2])</code></pre>
<pre><code>                                    1   3   4   5   6   7   8   9
BILE_ACID_METABOLISM              yes  no  no  no  no  no  no  no
MYC_TARGETS_V2                     no yes  no  no  no  no  no  no
E2F_TARGETS                        no  no yes  no  no  no  no  no
MYC_TARGETS_V1                     no  no  no yes  no  no  no  no
OXIDATIVE_PHOSPHORYLATION          no  no  no yes  no  no  no  no
MYOGENESIS                         no  no  no  no yes yes  no  no
MTORC1_SIGNALING                   no  no  no  no yes  no  no  no
HYPOXIA                            no  no  no  no yes  no  no yes
EPITHELIAL_MESENCHYMAL_TRANSITION  no  no  no  no yes yes yes yes
IL2_STAT5_SIGNALING                no  no  no  no yes  no  no  no
UV_RESPONSE_DN                     no  no  no  no yes  no  no yes
INFLAMMATORY_RESPONSE              no  no  no  no  no yes  no  no
TNFA_SIGNALING_VIA_NFKB            no  no  no  no  no yes  no yes
COAGULATION                        no  no  no  no  no  no  no yes
ANGIOGENESIS                       no  no  no  no  no  no  no yes</code></pre>
<pre class="r"><code>enriched_list &lt;- list()  # To store enrichment results

for (i in 1:num_clusters) {
  # For each cluster, perform enrichment and store the results
  enrich_res &lt;- enrich_cluster(cluster_id = i, q_val_cutoff = 1, melted_data = melted_data_quantiles)  # Set q_val_cutoff to 1 to ensure all p-values are returned
  
  # Store both the descriptions and p-values (not filtered by significance)
  enriched_list[[i]] &lt;- data.frame(
    Description = enrich_res$Description, 
    PValue = enrich_res$pvalue  # Use the raw p-value
  )
  
  names(enriched_list)[i] &lt;- paste0(i)
}

# Automatically create a list of all unique enriched terms
all_enriched_terms &lt;- unique(unlist(lapply(enriched_list, function(x) x$Description)))

# Remove the &quot;HALLMARK_&quot; prefix from each term
all_enriched_terms_clean &lt;- gsub(&quot;HALLMARK_&quot;, &quot;&quot;, all_enriched_terms)

# Initialize a data frame with clusters as columns and terms as rows
enrichment_table &lt;-
  data.frame(matrix(
    nrow = length(all_enriched_terms_clean),
    ncol = length(enriched_list)
  ))
rownames(enrichment_table) &lt;- all_enriched_terms_clean
colnames(enrichment_table) &lt;- names(enriched_list)

# Fill the table with p-values
for (cluster in names(enriched_list)) {
  for (term in all_enriched_terms_clean) {
    # Check if the term is present in the enriched list of the cluster
    original_term &lt;- paste0(&quot;HALLMARK_&quot;, term)  # Add back the prefix for comparison
    term_in_cluster &lt;- if (nrow(enriched_list[[cluster]]) &gt; 0) {
      enriched_list[[cluster]]$Description == original_term
    } else {
      FALSE
    }
    
    # If the term is enriched, fill in the p-value; otherwise, compute the enrichment and add its p-value
    if (any(term_in_cluster)) {
      enrichment_table[term, cluster] &lt;- round(enriched_list[[cluster]]$PValue[term_in_cluster], 3)
    } 
    else {
      # If not found in the original result, perform enrichment for the gene set specifically
      cluster_id &lt;- as.numeric(cluster)
      genes_selected &lt;- melted_data_quantiles %&gt;% filter(cluster == cluster_id) %&gt;% pull(id) %&gt;% as.character() %&gt;% unique()
      
      genes_converted &lt;- getBM(
        filters = &quot;ensembl_gene_id&quot;,
        attributes = c(&quot;ensembl_gene_id&quot;, &quot;entrezgene_id&quot;),
        values = genes_selected,
        mart = mart
      )
      
      # Extract Entrez IDs from the converted data
      entrez_gene_list &lt;- genes_converted$entrezgene_id
      
      genes_converted_all &lt;- getBM(
        filters = &quot;ensembl_gene_id&quot;,
        attributes = c(&quot;ensembl_gene_id&quot;, &quot;entrezgene_id&quot;),
        values = melted_data_quantiles$id %&gt;% as.character() %&gt;% unique(),
        mart = mart
      )
      entrez_universe &lt;- as.character(genes_converted_all$entrezgene_id)
      entrez_universe &lt;- entrez_universe[!is.na(entrez_universe)]
      
      enrich_res_full &lt;- enricher(entrez_gene_list,
                 TERM2GENE = m_t2g,
                 pvalueCutoff = 2, qvalueCutoff = 2,
                 universe = entrez_universe)
      
      
      specific_term &lt;- enrich_res_full %&gt;% filter(Description == original_term)
      
      # Add p-value if it exists
      if (nrow(specific_term) &gt; 0) {
        enrichment_table[term, cluster] &lt;- round(specific_term$p.adjust, 3)
      } else {
        enrichment_table[term, cluster] &lt;- 1  # No p-value found
      }
    }
  }
}

# Display the resulting table with p-values
print(enrichment_table)</code></pre>
</div>
<div
id="question-what-enriched-gene-sets-were-not-enriched-in-strobers-analysis"
class="section level3">
<h3><strong>Question: what enriched gene sets were not enriched in
Strober’s analysis? </strong></h3>
<p>Here are the enrichment result from Strober’s analysis for
comparison:</p>
<pre><code>                                    4   5  13  16  11   2   6  9
BILE_ACID_METABOLISM              yes  no  no  no  no  no  no no
MYC_TARGETS_V2                     no yes  no  no  no  no  no no
E2F_TARGETS                        no yes yes yes  no  no  no no
MYC_TARGETS_V1                     no yes yes yes  no  no  no no
OXIDATIVE_PHOSPHORYLATION          no  no  no  no yes  no  no no
MYOGENESIS                         no  no  no  no  no yes yes no
MTORC1_SIGNALING                   no yes  no  no  no  no  no no
HYPOXIA                            no  no  no  no  no  no  no no
EPITHELIAL_MESENCHYMAL_TRANSITION yes  no  no  no  no  no yes no
IL2_STAT5_SIGNALING                no  no  no  no  no  no  no no
UV_RESPONSE_DN                     no  no  no  no  no  no  no no
INFLAMMATORY_RESPONSE              no  no  no  no  no  no  no no
TNFA_SIGNALING_VIA_NFKB            no  no  no  no  no  no yes no
COAGULATION                       yes  no  no  no  no  no  no no
ANGIOGENESIS                       no  no  no  no  no  no  no no</code></pre>
<p>A few interesting observations:</p>
<ul>
<li><p>Many gene-sets are enriched in clusters identified from both
approaches. In particular, cluster 7 from FASH identified MYOGENESIS,
EPITHELIAL_MESENCHYMAL_TRANSITION, INFLAMMATORY_RESPONSE and
TNFA_SIGNALING_VIA_NFKB as enriched gene sets, whereas the cluster 6
from Strober identified MYOGENESIS, EPITHELIAL_MESENCHYMAL_TRANSITION
and TNFA_SIGNALING_VIA_NFKB.</p></li>
<li><p>Clusters (especially cluster 9) from FASH identifies some
interesting gene-sets that were missed by Strober’s analysis, such as
UV_RESPONSE_DN, INFLAMMATORY_RESPONSE, IL2_STAT5_SIGNALING, ANGIOGENESIS
and HYPOXIA. Note that UV_RESPONSE_DN INFLAMMATORY_RESPONSE and
IL2_STAT5_SIGNALING are all related to protective features to prevent
external damage due to environment or pathogens. ANGIOGENESIS and
HYPOXIA are related to the formation of new blood vessels and the
response to low oxygen levels, respectively.</p></li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
 [1] stats4    parallel  splines   stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] cowplot_1.1.3          biomaRt_2.58.2         org.Hs.eg.db_3.18.0   
 [4] AnnotationDbi_1.64.1   IRanges_2.36.0         S4Vectors_0.40.2      
 [7] Biobase_2.62.0         BiocGenerics_0.48.1    msigdbr_7.5.1         
[10] clusterProfiler_4.10.1 lubridate_1.9.3        forcats_1.0.0         
[13] stringr_1.5.1          dplyr_1.1.4            purrr_1.0.2           
[16] readr_2.1.5            tidyr_1.3.1            tibble_3.2.1          
[19] tidyverse_2.0.0        mixsqp_0.3-54          reshape2_1.4.4        
[22] ggplot2_3.5.1          Matrix_1.6-3           TMB_1.9.15            
[25] BayesGP_0.1.2          workflowr_1.7.1       

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      rstudioapi_0.16.0       jsonlite_1.8.9         
  [4] magrittr_2.0.3          farver_2.1.2            rmarkdown_2.28         
  [7] fs_1.6.4                zlibbioc_1.48.2         vctrs_0.6.5            
 [10] memoise_2.0.1           RCurl_1.98-1.16         ggtree_3.10.1          
 [13] progress_1.2.3          htmltools_0.5.8.1       curl_5.2.3             
 [16] gridGraphics_0.5-1      sass_0.4.9              bslib_0.8.0            
 [19] plyr_1.8.9              cachem_1.1.0            whisker_0.4.1          
 [22] igraph_2.0.3            lifecycle_1.0.4         pkgconfig_2.0.3        
 [25] gson_0.1.0              R6_2.5.1                fastmap_1.2.0          
 [28] GenomeInfoDbData_1.2.11 digest_0.6.37           aplot_0.2.3            
 [31] enrichplot_1.22.0       colorspace_2.1-1        patchwork_1.3.0        
 [34] ps_1.8.0                rprojroot_2.0.4         irlba_2.3.5.1          
 [37] RSQLite_2.3.7           filelock_1.0.3          labeling_0.4.3         
 [40] fansi_1.0.6             timechange_0.3.0        httr_1.4.7             
 [43] polyclip_1.10-7         compiler_4.3.1          bit64_4.5.2            
 [46] withr_3.0.1             BiocParallel_1.36.0     viridis_0.6.5          
 [49] DBI_1.2.3               highr_0.11              ggforce_0.4.2          
 [52] MASS_7.3-60             rappdirs_0.3.3          HDO.db_0.99.1          
 [55] tools_4.3.1             scatterpie_0.2.4        ape_5.8                
 [58] httpuv_1.6.15           glue_1.7.0              callr_3.7.6            
 [61] nlme_3.1-166            GOSemSim_2.28.1         promises_1.3.0         
 [64] shadowtext_0.1.4        grid_4.3.1              getPass_0.2-4          
 [67] fgsea_1.28.0            generics_0.1.3          gtable_0.3.5           
 [70] tzdb_0.4.0              data.table_1.16.0       hms_1.1.3              
 [73] xml2_1.3.6              tidygraph_1.3.1         utf8_1.2.4             
 [76] XVector_0.42.0          ggrepel_0.9.6           pillar_1.9.0           
 [79] babelgene_22.9          yulab.utils_0.1.7       later_1.3.2            
 [82] tweenr_2.0.3            BiocFileCache_2.10.2    treeio_1.26.0          
 [85] lattice_0.22-6          bit_4.5.0               tidyselect_1.2.1       
 [88] GO.db_3.18.0            Biostrings_2.70.3       knitr_1.48             
 [91] git2r_0.33.0            gridExtra_2.3           xfun_0.47              
 [94] graphlayouts_1.2.0      stringi_1.8.4           lazyeval_0.2.2         
 [97] ggfun_0.1.6             yaml_2.3.10             evaluate_1.0.0         
[100] codetools_0.2-20        ggraph_2.2.1            qvalue_2.34.0          
[103] ggplotify_0.1.2         cli_3.6.3               munsell_0.5.1          
[106] processx_3.8.4          jquerylib_0.1.4         Rcpp_1.0.13            
[109] GenomeInfoDb_1.38.8     dbplyr_2.5.0            png_0.1-8              
[112] XML_3.99-0.17           blob_1.2.4              prettyunits_1.2.0      
[115] DOSE_3.28.2             bitops_1.0-8            tidytree_0.4.6         
[118] viridisLite_0.4.2       scales_1.3.0            crayon_1.5.3           
[121] rlang_1.1.4             fastmatch_1.1-4         KEGGREST_1.42.0        </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
