<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Example: Expression Analysis</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">resultsummary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="simulation.html">Simulation 1</a>
</li>
<li>
  <a href="simulation_eQTL_identification.html">Simulation 2</a>
</li>
<li>
  <a href="covid_example.html">COVID-19 Example</a>
</li>
<li>
  <a href="expression.html">Gene Expression Example</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/AgueroZZ/resultsummary">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Example: Expression Analysis</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-05-27
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>FASHresultsummary/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240507code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240507)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240507code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240507)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytree2aadd9258476fc0b8b8398684ed6c962a840c5e0targetblank2aadd92a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/2aadd9258476fc0b8b8398684ed6c962a840c5e0" target="_blank">2aadd92</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytree2aadd9258476fc0b8b8398684ed6c962a840c5e0targetblank2aadd92a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/2aadd9258476fc0b8b8398684ed6c962a840c5e0" target="_blank">2aadd92</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    code/.DS_Store
    Ignored:    code/cpp/.DS_Store
    Ignored:    code/function/.DS_Store
    Ignored:    code/function/.Rhistory
    Ignored:    data/.DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/example/.DS_Store
    Ignored:    output/example/figure/
    Ignored:    output/simulation/.DS_Store
    Ignored:    output/simulation/figure/

Untracked files:
    Untracked:  code/cpp/Gaussian_just_fixed.o
    Untracked:  code/cpp/Gaussian_just_fixed.so
    Untracked:  code/cpp/Gaussian_theta_known.o
    Untracked:  code/cpp/Gaussian_theta_known.so
    Untracked:  code/cpp/Poisson.o
    Untracked:  code/cpp/Poisson.so
    Untracked:  code/cpp/Poisson_causes.o
    Untracked:  code/cpp/Poisson_causes.so
    Untracked:  code/cpp/Poisson_covid.o
    Untracked:  code/cpp/Poisson_covid.so
    Untracked:  code/cpp/Poisson_just_fixed.cpp
    Untracked:  code/cpp/Poisson_just_fixed.o
    Untracked:  code/cpp/Poisson_just_fixed.so
    Untracked:  code/cpp/Poisson_just_fixed_causes.o
    Untracked:  code/cpp/Poisson_just_fixed_causes.so
    Untracked:  code/cpp/Poisson_just_fixed_covid.o
    Untracked:  code/cpp/Poisson_just_fixed_covid.so
    Untracked:  code/function/functions_fitting_Gaussian_eQTL_irregular.R
    Untracked:  code/function/functions_simulation_eQTL_irregular.R
    Untracked:  data/13100810.csv
    Untracked:  data/13100810_MetaData.csv
    Untracked:  data/expression_data/
    Untracked:  exp.R
    Untracked:  output/causes/
    Untracked:  output/expression/
    Untracked:  output/simulation_eQTL/
    Untracked:  output/simulation_eQTL_test/

Unstaged changes:
    Modified:   output/example/L_matrix.rda
    Modified:   output/example/fitted_datasets.rda
    Modified:   output/example/functional_datasets.rda
    Modified:   output/example/posterior_weights_matrix.rda

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/expression.Rmd</code>) and HTML
(<code>docs/expression.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/ab342cbde768b4282e6db6a241018b33444b998a/analysis/expression.Rmd" target="_blank">ab342cb</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-27
</td>
<td>
wflow_git_commit(c("analysis/simulation_eQTL_identification.Rmd",
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/3b37c543448424223e37bef6f636d8c9dce217ec/docs/expression.html" target="_blank">3b37c54</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
update html
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/08cc7a7aebe5d38aa45f16a4164627dc04555198/analysis/expression.Rmd" target="_blank">08cc7a7</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
wflow_git_commit(c("analysis/expression.Rmd",
"analysis/covid_example.Rmd"))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/expression.html" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
update the expression example
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/f571ee01ac5a385ecb0d9a7678a133062d970451/analysis/expression.Rmd" target="_blank">f571ee0</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
<td>
wflow_git_commit(c("analysis/covid_example.Rmd",
"code/function/functions_fitting_Poisson_covid.R",
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="setup" class="section level3">
<h3><strong>Setup</strong></h3>
<p>We consider the (raw) gene expression data measured over 16 days as
studied in <a
href="https://www.science.org/doi/10.1126/science.aaw0040">Strober et
al, 2019</a>. For simplicity of the presentation, we consider only one
cell-line in this example. The dataset contains the raw expression
counts of 16319 genes.</p>
<p>To model the true (log) expression level <span
class="math inline">\(f_i(t_j)\)</span> of the <span
class="math inline">\(i\)</span>th gene at day <span
class="math inline">\(t_j\)</span>, we assume that the expression count
<span class="math inline">\(y_{i}(t_j)\)</span> is Poisson distributed
with mean <span class="math inline">\(\exp
\left(f_i(t_j)\right)\)</span>.</p>
</div>
<div id="model" class="section level3">
<h3><strong>Model</strong></h3>
<p>Assume most of the gene expression trajectories are expected to be
linear, we therefore consider the following finite mixture prior for
each gene expression level <span
class="math inline">\(f_i(t_j)\)</span>: <span
class="math display">\[f_i(t) = \sum_{k=1}^K \pi_k g_k(t).\]</span> Each
mixture component <span class="math inline">\(g_k\)</span> has the
prior: <span class="math display">\[Lg_k(t) = \sigma_k W(t), \]</span>
with <span class="math inline">\(W(t)\)</span> being a Gaussian white
noise process and <span class="math inline">\(\sigma_k\)</span> being
the smoothness. The base model that each <span
class="math inline">\(f_i\)</span> being shrunk to is therefore <span
class="math inline">\(\text{Null}\{L\} =
\text{span}\{1,t\}\)</span>.</p>
</div>
<div id="processing-the-expression-data" class="section level3">
<h3><strong>Processing the expression data</strong></h3>
<pre class="r"><code>all_data_names &lt;- list.files(data_dir)
#### let&#39;s consider one cell-line for simplicity
load(paste0(data_dir, &quot;/&quot;, all_data_names[1]))
str(expression_data_one_cell_line) </code></pre>
<pre><code>tibble [261,104 × 5] (S3: tbl_df/tbl/data.frame)
 $ Gene_id           : chr [1:261104] &quot;ENSG00000000419&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000419&quot; ...
 $ Cell_line         : chr [1:261104] &quot;18489&quot; &quot;18489&quot; &quot;18489&quot; &quot;18489&quot; ...
 $ Day               : int [1:261104] 0 1 2 3 4 5 6 7 8 9 ...
 $ value             : int [1:261104] 825 2432 1044 874 766 632 839 685 640 636 ...
 $ external_gene_name: chr [1:261104] &quot;DPM1&quot; &quot;DPM1&quot; &quot;DPM1&quot; &quot;DPM1&quot; ...</code></pre>
<p>Let’s create one dataset for each gene:</p>
<pre class="r"><code>num_knots &lt;- 8
p &lt;- 2
datasets &lt;- list()
all_gene &lt;- unique(expression_data_one_cell_line$Gene_id)
for (gene in all_gene) {
  datasets[[gene]] &lt;- expression_data_one_cell_line %&gt;% filter(Gene_id == gene)
  datasets[[gene]]$x &lt;- as.numeric(datasets[[gene]]$Day)
  datasets[[gene]]$y &lt;- datasets[[gene]]$value
}
save(datasets, file = paste0(result_dir, &quot;/datasets.rda&quot;))</code></pre>
</div>
<div id="applying-empirical-bayes" class="section level3">
<h3><strong>Applying empirical Bayes</strong></h3>
<pre class="r"><code>log_prec &lt;- seq(0,5, length.out = 10)
psd_vec &lt;- sort(c(0,exp(-.5*log_prec)))

## set up a progress bar
L_vecs &lt;- vector(&quot;list&quot;, length(datasets))
pb &lt;- txtProgressBar(min = 0, max = length(datasets), style = 3)
for (i in 1:length(datasets)) {
  setTxtProgressBar(pb, i)
  dataset &lt;- datasets[[i]]
  L_vecs[[i]] &lt;- compute_log_likelihood_ospline_seq2(
    dataset = dataset,
    p = p,
    num_knots = num_knots,
    psd_iwp_vector = psd_vec,
    pred_step = 1,
    betaprec = 0.0001
  )
}
L_matrix &lt;- do.call(rbind, L_vecs)
save(L_matrix, file = paste0(result_dir, &quot;/L_matrix.rda&quot;))</code></pre>
<pre class="r"><code>fit.sqp &lt;- mixsqp(L = L_matrix, log = TRUE)</code></pre>
<pre><code>Running mix-SQP algorithm 0.3-54 on 16319 x 11 matrix
convergence tol. (SQP):     1.0e-08
conv. tol. (active-set):    1.0e-10
zero threshold (solution):  1.0e-08
zero thresh. (search dir.): 1.0e-14
l.s. sufficient decrease:   1.0e-02
step size reduction factor: 7.5e-01
minimum step size:          1.0e-08
max. iter (SQP):            1000
max. iter (active-set):     12
number of EM iterations:    10
Computing SVD of 16319 x 11 matrix.
Matrix is not low-rank; falling back to full matrix.
iter        objective max(rdual) nnz stepsize max.diff nqp nls
   1 +1.470056684e+00  -- EM --   11 1.00e+00 7.54e-02  --  --
   2 +1.460919861e+00  -- EM --   11 1.00e+00 1.70e-02  --  --
   3 +1.458341741e+00  -- EM --   11 1.00e+00 9.25e-03  --  --
   4 +1.457322915e+00  -- EM --   11 1.00e+00 5.49e-03  --  --
   5 +1.456847151e+00  -- EM --   11 1.00e+00 3.49e-03  --  --
   6 +1.456587509e+00  -- EM --   11 1.00e+00 2.32e-03  --  --
   7 +1.456421502e+00  -- EM --   11 1.00e+00 1.61e-03  --  --
   8 +1.456300041e+00  -- EM --   11 1.00e+00 1.16e-03  --  --
   9 +1.456202478e+00  -- EM --   11 1.00e+00 1.01e-03  --  --
  10 +1.456119628e+00  -- EM --   11 1.00e+00 9.05e-04  --  --
   1 +1.456047074e+00 +8.791e-03  11  ------   ------   --  --
   2 +1.455124498e+00 +6.637e-04  10 1.00e+00 5.01e-02   3   1
   3 +1.455124420e+00 +2.532e-07  10 1.00e+00 6.55e-04   2   1
   4 +1.455124420e+00 -3.590e-08  10 1.00e+00 4.48e-07   2   1
Optimization took 0.02 seconds.
Convergence criteria met---optimal solution found.</code></pre>
<pre class="r"><code>numiter &lt;- nrow(fit.sqp$progress)
plot(1:numiter,fit.sqp$progress$objective,type = &quot;b&quot;,
     pch = 20,lwd = 2,xlab = &quot;SQP iteration&quot;,
     ylab = &quot;objective&quot;,xaxp = c(1,numiter,numiter - 1))</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-7-1">
Past versions of unnamed-chunk-7-1.png
</button>
</p>
<div id="fig-unnamed-chunk-7-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-7-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>prior_weight &lt;- data.frame(p = rep(p, each = length(psd_vec)), psd_iwp = psd_vec, prior_weight = fit.sqp$x)</code></pre>
</div>
<div id="obtaining-posteriors" class="section level3">
<h3><strong>Obtaining posteriors</strong></h3>
<pre class="r"><code>num_datasets &lt;- length(datasets)
num_weights &lt;- sum(prior_weight$prior_weight != 0)
posterior_weights_matrix &lt;- matrix(nrow = num_datasets, ncol = num_weights)

# Loop through each dataset and perform fitting
fitted_datasets &lt;- list()
# create a progress bar
pb &lt;- txtProgressBar(min = 0, max = num_datasets, style = 3)
for (i in seq_along(datasets)) {
  setTxtProgressBar(pb, i)
  dataset &lt;- datasets[[i]]
  fit_result_final &lt;- fit_ospline_with_prior2(
    num_cores = num_cores,
    dataset = dataset,
    num_knots = num_knots,
    prior_weight = prior_weight,
    betaprec = 0.0001,
    pred_step = 1
  )
  posterior_weights_matrix[i, ] &lt;- fit_result_final$posterior_weights[, &quot;posterior_weight&quot;]
  fitted_datasets[[i]] &lt;- aggregate_fit_with_prior(x = dataset$x, fit_results_with_prior = fit_result_final, original = TRUE)$summary_df
}
names(fitted_datasets) &lt;- all_gene
colnames(posterior_weights_matrix) &lt;- paste(as.character(fit_result_final$posterior_weights[, &quot;p&quot;]),
                                            as.character(fit_result_final$posterior_weights[, &quot;psd_iwp&quot;]), sep = &quot;_&quot;)
save(fitted_datasets, posterior_weights_matrix, file = paste0(result_dir, &quot;/fitted_pos_results.rda&quot;))</code></pre>
<pre class="r"><code>load(paste0(result_dir, &quot;/fitted_pos_results.rda&quot;))</code></pre>
<pre class="r"><code>posterior_weights_df &lt;- as.data.frame(posterior_weights_matrix)
posterior_weights_df$id &lt;- all_gene
melted_data &lt;- melt(posterior_weights_df, id.vars = &quot;id&quot;)
melted_data$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data$variable)
melted_data$variable3 &lt;- (round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data$variable)), 3))

ggplot(melted_data, aes(x = id, y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;, width = 0.7) +  # Adjust bar width if necessary
  labs(x = &quot;Gene&quot;, y = &quot;Weight&quot;, fill = &quot;PSD&quot;) +
  ggtitle(&quot;Structure Plot of Posterior Weights&quot;) +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  coord_flip() +  # Flips the coordinates to make &#39;Gene&#39; on the y-axis
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),  
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA) 
  )</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-10-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot a sorted version of this, based on the mean PSD in each
gene:</p>
<pre class="r"><code>mean_psd &lt;- apply(posterior_weights_matrix, 1, function(x) {
  weighted_psd &lt;- as.numeric(sub(&quot;*._&quot;, &quot;&quot;, colnames(posterior_weights_matrix)))
  sum(x * weighted_psd)
})
sorted_index &lt;- order(mean_psd)
sorted_posterior_weights_matrix &lt;- posterior_weights_matrix[sorted_index, ]
sorted_posterior_weights_df &lt;- as.data.frame(sorted_posterior_weights_matrix)
sorted_posterior_weights_df$id &lt;- all_gene[sorted_index]
melted_data_sorted &lt;- melt(sorted_posterior_weights_df, id.vars = &quot;id&quot;)
melted_data_sorted$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data_sorted$variable)
melted_data_sorted$variable3 &lt;- (round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data_sorted$variable)), 3))

ggplot(melted_data_sorted, aes(x = factor(id, levels = sorted_posterior_weights_df$id), y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +
  labs(x = &quot;Gene&quot;, y = &quot;Weight&quot;, fill = &quot;PSD&quot;) +
  ggtitle(&quot;Structure Plot of Posterior Weights&quot;) +
  coord_flip() +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),  
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA) 
  )</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-11-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="clustering-based-on-posterior-weights" class="section level3">
<h3><strong>Clustering based on posterior weights</strong></h3>
<p>Let’s group them into 6 groups, first try hierarchical clustering</p>
<pre class="r"><code>set.seed(123)
# Normalize the data
wide_data &lt;- scale(sorted_posterior_weights_matrix)
# Perform hierarchical clustering
d &lt;- dist(wide_data, method = &quot;euclidean&quot;)  # Distance matrix
fit &lt;- hclust(d, method = &quot;ward.D2&quot;)  # Clustering
clusters &lt;- cutree(fit, k = 6)
# recode cluster 1 - 1, 2 - 2, 3 - 3, 5 - 4, 6 - 5, 4 - 6
clusters &lt;- recode(clusters, `1` = 1, `2` = 2, `3` = 3, `4` = 5, `5` = 4, `6` = 6)

sorted_posterior_weights_df$cluster &lt;- clusters
melted_data_sorted &lt;- melt(sorted_posterior_weights_df, id.vars = c(&quot;id&quot;, &quot;cluster&quot;))
melted_data_sorted$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data_sorted$variable)
melted_data_sorted$variable3 &lt;- (round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data_sorted$variable)), 3))
melted_data_sorted &lt;- melted_data_sorted %&gt;% arrange(cluster)

ggplot(melted_data_sorted, aes(x = factor(id), y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +  
  labs(x = &quot;Gene Index&quot;, y = &quot;Weight&quot;, fill = &quot;PSD Level&quot;) +
  ggtitle(&quot;Structure Plot of Posterior Weights by Cluster&quot;) +
  coord_flip() +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  facet_wrap(~cluster, scales = &quot;free_y&quot;) +  # Use facet_wrap by cluster
  theme_minimal() +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA),
    strip.background = element_rect(fill = &quot;lightblue&quot;, colour = &quot;deepskyblue&quot;, size = 1),
    strip.text = element_text(face = &quot;bold&quot;)  
  )</code></pre>
<pre><code>Warning: The `size` argument of `element_rect()` is deprecated as of ggplot2 3.4.0.
ℹ Please use the `linewidth` argument instead.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-1">
Past versions of unnamed-chunk-12-1.png
</button>
</p>
<div id="fig-unnamed-chunk-12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-12-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Then try to group just based on the quantiles of the mean PSD:</p>
<pre class="r"><code>mean_psd_ordered &lt;- mean_psd[sorted_index]
quantiles &lt;- quantile(mean_psd_ordered, probs = seq(0, 1, 1/6))
quantile_groups &lt;- cut(mean_psd_ordered, breaks = quantiles, 
                       include.lowest = TRUE, labels = FALSE)

sorted_posterior_weights_df$cluster &lt;- quantile_groups
melted_data_quantiles &lt;- melt(sorted_posterior_weights_df, id.vars = c(&quot;id&quot;, &quot;cluster&quot;))
melted_data_quantiles$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data_quantiles$variable)
melted_data_quantiles$variable3 &lt;- round(as.numeric(sub(&quot;.*_&quot;, &quot;&quot;, melted_data_quantiles$variable)), 3)
melted_data_quantiles &lt;- melted_data_quantiles %&gt;% arrange(cluster)

# Plotting
ggplot(melted_data_quantiles, aes(x = factor(id), y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +
  labs(x = &quot;Gene Index&quot;, y = &quot;Weight&quot;, fill = &quot;PSD Level&quot;) +
  ggtitle(&quot;Structure Plot of Posterior Weights by Quantile Group&quot;) +
  coord_flip() +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  facet_wrap(~cluster, scales = &quot;free_y&quot;) +
  theme_minimal() +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA),
    strip.background = element_rect(fill = &quot;lightblue&quot;, colour = &quot;deepskyblue&quot;, size = 1),
    strip.text = element_text(face = &quot;bold&quot;)
  )</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-13-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>plot_gene_expression &lt;- function(gene) {
  agg_result &lt;- fitted_datasets[[gene]]
  plot(datasets[[gene]]$Day, agg_result$median, type = &#39;l&#39;, 
       main = paste0(gene), xlab = &quot;Day&quot;, ylab = &quot;Expression&quot;, col = &quot;blue&quot;,
       cex = 0.5, ylim = c(0, max(datasets[[gene]]$y) * 1.2),
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)
  points(datasets[[gene]]$Day, datasets[[gene]]$y, col = &quot;black&quot;, cex = 0.5)
  polygon(c(datasets[[gene]]$Day, rev(datasets[[gene]]$Day)),
          c(agg_result$lower, rev(agg_result$upper)),
          col = rgb(0.6, 0.8, 1, alpha = 0.3), border = NA)
}

plot_gene_expression_log &lt;- function(gene) {
  agg_result &lt;- fitted_datasets[[gene]]
  data_points &lt;- datasets[[gene]]$y
  positive_points &lt;- data_points[data_points &gt; 0]  # Filter out non-positive values

  if (length(positive_points) == 0) {
    stop(&quot;No positive data points available for logarithmic plotting.&quot;)
  }

  min_y &lt;- min(positive_points)
  max_y &lt;- max(agg_result$upper)  # Assuming this is positive and makes sense in your context

  # Set the limits for the y-axis
  log_min_y &lt;- min_y / 2  # Just a simple heuristic, can adjust based on your data&#39;s distribution
  log_max_y &lt;- max_y * 1.2

  plot(datasets[[gene]]$Day, agg_result$median, type = &#39;l&#39;, 
       main = paste0(gene), xlab = &quot;Day&quot;, ylab = &quot;Expression&quot;, col = &quot;blue&quot;,
       cex = 0.5, log = &quot;y&quot;, ylim = c(log_min_y, log_max_y),
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)
  points(datasets[[gene]]$Day, data_points, col = &quot;black&quot;, cex = 0.5)
  polygon(c(datasets[[gene]]$Day, rev(datasets[[gene]]$Day)),
          c(agg_result$lower, rev(agg_result$upper)),
          col = rgb(0.6, 0.8, 1, alpha = 0.3), border = NA)
}

# plot_gene_expression(gene)
plot_cluster_genes &lt;- function(genes) {
  if(length(genes) &lt; 16) stop(&quot;Not enough genes in the cluster&quot;)
  par(mfrow = c(4, 4), mar = c(4, 4, 2, 1))  # Adjust margins if necessary
  for (gene in genes[1:16]) {
    plot_gene_expression(gene)
  }
  par(mfrow = c(1, 1))
}
plot_cluster_genes_log &lt;- function(genes) {
  if(length(genes) &lt; 16) stop(&quot;Not enough genes in the cluster&quot;)
  par(mfrow = c(4, 4), mar = c(4, 4, 2, 1))  # Adjust margins if necessary
  for (gene in genes[1:16]) {
    plot_gene_expression_log(gene)
  }
  par(mfrow = c(1, 1))
}
# plot_cluster_genes(genes_cluster1)</code></pre>
<pre class="r"><code>num_genes_per_cluster &lt;- 16
num_clusters &lt;- 6
genes_cluster1 &lt;- sorted_posterior_weights_df$id[sorted_posterior_weights_df$cluster == 1][1:num_genes_per_cluster]
genes_cluster2 &lt;- sorted_posterior_weights_df$id[sorted_posterior_weights_df$cluster == 2][1:num_genes_per_cluster]
genes_cluster3 &lt;- sorted_posterior_weights_df$id[sorted_posterior_weights_df$cluster == 3][1:num_genes_per_cluster]
genes_cluster4 &lt;- sorted_posterior_weights_df$id[sorted_posterior_weights_df$cluster == 4][1:num_genes_per_cluster]
genes_cluster5 &lt;- sorted_posterior_weights_df$id[sorted_posterior_weights_df$cluster == 5][1:num_genes_per_cluster]
genes_cluster6 &lt;- sorted_posterior_weights_df$id[sorted_posterior_weights_df$cluster == 6][1:num_genes_per_cluster]</code></pre>
</div>
<div id="smoothing-result-for-each-cluster"
class="section level3 tabset tabset-pills">
<h3 class="tabset tabset-pills"><strong>Smoothing result for each
cluster</strong></h3>
<div id="for-cluster-1" class="section level4">
<h4>For cluster 1:</h4>
<pre class="r"><code>plot_cluster_genes(genes_cluster1)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-16-1">
Past versions of unnamed-chunk-16-1.png
</button>
</p>
<div id="fig-unnamed-chunk-16-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-16-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-2" class="section level4">
<h4>For cluster 2:</h4>
<pre class="r"><code>plot_cluster_genes(genes_cluster2)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-17-1">
Past versions of unnamed-chunk-17-1.png
</button>
</p>
<div id="fig-unnamed-chunk-17-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-17-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-3" class="section level4">
<h4>For cluster 3:</h4>
<pre class="r"><code>plot_cluster_genes(genes_cluster3)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-18-1">
Past versions of unnamed-chunk-18-1.png
</button>
</p>
<div id="fig-unnamed-chunk-18-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-18-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-4" class="section level4">
<h4>For cluster 4:</h4>
<pre class="r"><code>plot_cluster_genes(genes_cluster4)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-19-1">
Past versions of unnamed-chunk-19-1.png
</button>
</p>
<div id="fig-unnamed-chunk-19-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-19-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-5" class="section level4">
<h4>For cluster 5:</h4>
<pre class="r"><code>plot_cluster_genes(genes_cluster5)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-20-1">
Past versions of unnamed-chunk-20-1.png
</button>
</p>
<div id="fig-unnamed-chunk-20-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-20-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-6" class="section level4">
<h4>For cluster 6:</h4>
<pre class="r"><code>plot_cluster_genes(genes_cluster6)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-21-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="smoothing-result-for-each-cluster-after-log"
class="section level3 tabset tabset-pills">
<h3 class="tabset tabset-pills"><strong>Smoothing result for each
cluster (after Log)</strong></h3>
<div id="for-cluster-1-1" class="section level4">
<h4>For cluster 1:</h4>
<pre class="r"><code>plot_cluster_genes_log(genes_cluster1)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-22-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-2-1" class="section level4">
<h4>For cluster 2:</h4>
<pre class="r"><code>plot_cluster_genes_log(genes_cluster2)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-23-1">
Past versions of unnamed-chunk-23-1.png
</button>
</p>
<div id="fig-unnamed-chunk-23-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-23-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-3-1" class="section level4">
<h4>For cluster 3:</h4>
<pre class="r"><code>plot_cluster_genes_log(genes_cluster3)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-24-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-4-1" class="section level4">
<h4>For cluster 4:</h4>
<pre class="r"><code>plot_cluster_genes_log(genes_cluster4)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-25-1">
Past versions of unnamed-chunk-25-1.png
</button>
</p>
<div id="fig-unnamed-chunk-25-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-25-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-5-1" class="section level4">
<h4>For cluster 5:</h4>
<pre class="r"><code>plot_cluster_genes_log(genes_cluster5)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-26-1">
Past versions of unnamed-chunk-26-1.png
</button>
</p>
<div id="fig-unnamed-chunk-26-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-26-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="for-cluster-6-1" class="section level4">
<h4>For cluster 6:</h4>
<pre class="r"><code>plot_cluster_genes_log(genes_cluster6)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-27-1">
Past versions of unnamed-chunk-27-1.png
</button>
</p>
<div id="fig-unnamed-chunk-27-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-27-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="question-what-genes-have-linear-trajectories"
class="section level3">
<h3><strong>Question: what genes have linear trajectories?
</strong></h3>
<pre class="r"><code>p0 &lt;- posterior_weights_df$`2_0` # prob of PSD = 0 -&gt; linear
local_fda &lt;- 1 - p0
fda_df &lt;- data.frame(gene = all_gene, fda = local_fda)
fda_df &lt;- fda_df[order(fda_df$fda), ] # ordering it
fda_df$cumulative_fda &lt;- cumsum(fda_df$fda)/seq_along(fda_df$fda)
ensembl &lt;- biomaRt::useMart(&quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)
# Assuming your data frame is &#39;gene_exp_ordered&#39; and it has a column &#39;Gene_id&#39;
gene_symbols &lt;- biomaRt::getBM(attributes = c(&#39;ensembl_gene_id&#39;, &#39;external_gene_name&#39;), 
                      filters = &#39;ensembl_gene_id&#39;, 
                      values = fda_df$gene, 
                      mart = ensembl)
colnames(gene_symbols)[1] &lt;- &quot;gene&quot;
colnames(gene_symbols)[2] &lt;- &quot;gene_name&quot;
fda_df &lt;- merge(fda_df, gene_symbols, by = &quot;gene&quot;) %&gt;% arrange(cumulative_fda)</code></pre>
<pre class="r"><code>plot(fda_df$cumulative_fda, type = &quot;l&quot;, xlab = &quot;Ordered Genes&quot;, ylab = &quot;Cumulative FDA&quot;, ylim = c(0,1), cex = 0.1)
abline(h = 0.05, col = &quot;purple&quot;, lty = &quot;dashed&quot;)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-29-1">
Past versions of unnamed-chunk-29-1.png
</button>
</p>
<div id="fig-unnamed-chunk-29-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-29-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>fda_df %&gt;% filter(cumulative_fda &lt; 0.05, !is.na(gene_name)) %&gt;% 
  arrange(cumulative_fda) %&gt;%
  head(16)</code></pre>
<pre><code>              gene          fda cumulative_fda gene_name
1  ENSG00000156958 5.272845e-08   5.272845e-08     GALK2
2  ENSG00000121542 9.561169e-08   7.417007e-08    SEC22A
3  ENSG00000009780 1.478928e-07   9.874430e-08    FAM76A
4  ENSG00000235109 1.484049e-07   1.111594e-07   ZSCAN31
5  ENSG00000196436 1.658680e-07   1.221012e-07   NPIPB15
6  ENSG00000172594 2.180770e-07   1.380971e-07   SMPDL3A
7  ENSG00000122696 2.337421e-07   1.517607e-07  SLC25A51
8  ENSG00000228696 2.339847e-07   1.620387e-07    ARL17B
9  ENSG00000240053 2.535231e-07   1.722036e-07    LY6G5B
10 ENSG00000169087 4.362178e-07   1.986051e-07   HSPBAP1
11 ENSG00000148950 6.979862e-07   2.440033e-07    IMMP1L
12 ENSG00000139537 7.395559e-07   2.852994e-07    CCDC65
13 ENSG00000186010 8.556404e-07   3.291718e-07   NDUFA13
14 ENSG00000171984 8.579557e-07   3.669421e-07     SHLD1
15 ENSG00000111215 8.684411e-07   4.003753e-07      PRR4
16 ENSG00000176623 8.815372e-07   4.304479e-07     RMDN1</code></pre>
<p>Plot the log expression of the top 16 genes:</p>
<pre class="r"><code>top_genes &lt;- fda_df %&gt;% filter(cumulative_fda &lt; 0.05, !is.na(gene_name)) %&gt;% 
  arrange(cumulative_fda) %&gt;%
  head(16) %&gt;% pull(gene)
plot_cluster_genes_log(top_genes)</code></pre>
<p><img src="figure/expression.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-31-1">
Past versions of unnamed-chunk-31-1.png
</button>
</p>
<div id="fig-unnamed-chunk-31-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/5f486a45d51713eaa0960b9dd010f9d13cf05ff5/docs/figure/expression.Rmd/unnamed-chunk-31-1.png" target="_blank">5f486a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Toronto
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.0   dplyr_1.1.3    
 [5] purrr_1.0.2     readr_2.1.4     tidyr_1.3.0     tibble_3.2.1   
 [9] tidyverse_2.0.0 mixsqp_0.3-54   reshape2_1.4.4  ggplot2_3.5.1  
[13] Matrix_1.6-3    TMB_1.9.7       BayesGP_0.1.1   workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0        farver_2.1.1            blob_1.2.4             
 [4] filelock_1.0.3          bitops_1.0-7            Biostrings_2.70.3      
 [7] RCurl_1.98-1.12         fastmap_1.1.1           BiocFileCache_2.10.2   
[10] promises_1.2.1          XML_3.99-0.14           digest_0.6.33          
[13] timechange_0.2.0        lifecycle_1.0.3         processx_3.8.2         
[16] KEGGREST_1.42.0         RSQLite_2.3.6           magrittr_2.0.3         
[19] compiler_4.3.1          progress_1.2.2          rlang_1.1.1            
[22] sass_0.4.7              tools_4.3.1             utf8_1.2.3             
[25] yaml_2.3.7              knitr_1.44              prettyunits_1.2.0      
[28] labeling_0.4.3          curl_5.1.0              bit_4.0.5              
[31] xml2_1.3.5              plyr_1.8.9              withr_2.5.1            
[34] BiocGenerics_0.48.1     grid_4.3.1              stats4_4.3.1           
[37] fansi_1.0.5             git2r_0.33.0            colorspace_2.1-0       
[40] scales_1.3.0            biomaRt_2.58.2          cli_3.6.1              
[43] rmarkdown_2.25          crayon_1.5.2            generics_0.1.3         
[46] rstudioapi_0.15.0       httr_1.4.7              tzdb_0.4.0             
[49] DBI_1.2.2               cachem_1.0.8            zlibbioc_1.48.2        
[52] AnnotationDbi_1.64.1    XVector_0.42.0          vctrs_0.6.4            
[55] jsonlite_1.8.7          callr_3.7.3             IRanges_2.36.0         
[58] hms_1.1.3               S4Vectors_0.40.2        bit64_4.0.5            
[61] irlba_2.3.5.1           jquerylib_0.1.4         glue_1.6.2             
[64] ps_1.7.5                stringi_1.7.12          gtable_0.3.4           
[67] GenomeInfoDb_1.38.8     later_1.3.1             munsell_0.5.0          
[70] pillar_1.9.0            rappdirs_0.3.3          htmltools_0.5.6.1      
[73] GenomeInfoDbData_1.2.11 dbplyr_2.3.4            R6_2.5.1               
[76] rprojroot_2.0.3         evaluate_0.22           lattice_0.21-8         
[79] Biobase_2.62.0          png_0.1-8               memoise_2.0.1          
[82] httpuv_1.6.11           bslib_0.5.1             Rcpp_1.0.11            
[85] whisker_0.4.1           xfun_0.40               fs_1.6.3               
[88] getPass_0.2-4           pkgconfig_2.0.3        </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
