<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Simulation: Pollution Measurement</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">resultsummary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="simulation_pollution.html">Simulation 1</a>
</li>
<li>
  <a href="simulation_eQTL_identification.html">Simulation 2</a>
</li>
<li>
  <a href="covid_example.html">COVID-19 Example</a>
</li>
<li>
  <a href="expression.html">Gene Expression Example</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/AgueroZZ/resultsummary">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Simulation: Pollution Measurement</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-09-05
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>FASHresultsummary/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240507code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240507)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240507code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240507)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreeb729c3febd0f9dc9309871c6d9ce8544ecff7d5etargetblankb729c3fa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/b729c3febd0f9dc9309871c6d9ce8544ecff7d5e" target="_blank">b729c3f</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreeb729c3febd0f9dc9309871c6d9ce8544ecff7d5etargetblankb729c3fa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/b729c3febd0f9dc9309871c6d9ce8544ecff7d5e" target="_blank">b729c3f</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    code/.DS_Store
    Ignored:    code/cpp/.DS_Store
    Ignored:    code/function/.DS_Store
    Ignored:    code/function/.Rhistory
    Ignored:    data/.DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/example/.DS_Store
    Ignored:    output/example/figure/

Untracked files:
    Untracked:  analysis/circadian_expression.Rmd
    Untracked:  code/cpp/Gaussian_just_fixed.o
    Untracked:  code/cpp/Gaussian_just_fixed.so
    Untracked:  code/cpp/Gaussian_theta_known.o
    Untracked:  code/cpp/Gaussian_theta_known.so
    Untracked:  code/cpp/Poisson.o
    Untracked:  code/cpp/Poisson.so
    Untracked:  code/cpp/Poisson_causes.o
    Untracked:  code/cpp/Poisson_causes.so
    Untracked:  code/cpp/Poisson_covid.o
    Untracked:  code/cpp/Poisson_covid.so
    Untracked:  code/cpp/Poisson_expression.o
    Untracked:  code/cpp/Poisson_expression.so
    Untracked:  code/cpp/Poisson_just_fixed.cpp
    Untracked:  code/cpp/Poisson_just_fixed.o
    Untracked:  code/cpp/Poisson_just_fixed.so
    Untracked:  code/cpp/Poisson_just_fixed_causes.o
    Untracked:  code/cpp/Poisson_just_fixed_causes.so
    Untracked:  code/cpp/Poisson_just_fixed_covid.o
    Untracked:  code/cpp/Poisson_just_fixed_covid.so
    Untracked:  code/cpp/Poisson_just_fixed_expression.o
    Untracked:  code/cpp/Poisson_just_fixed_expression.so
    Untracked:  code/function/functions_fitting_Gaussian_eQTL_irregular.R
    Untracked:  code/function/functions_fitting_Poisson_circadian_expression.R
    Untracked:  code/function/functions_simulation_eQTL_irregular.R
    Untracked:  data/13100810.csv
    Untracked:  data/13100810_MetaData.csv
    Untracked:  data/circadian_expression/
    Untracked:  data/expression_data/
    Untracked:  output/causes/
    Untracked:  output/circadian_expression/
    Untracked:  output/expression/
    Untracked:  output/simulation_eQTL/
    Untracked:  output/simulation_eQTL_test/
    Untracked:  output/simulation_pollution/

Unstaged changes:
    Deleted:    analysis/simulation.Rmd
    Modified:   analysis/simulation_eQTL.Rmd
    Modified:   analysis/simulation_pollution.Rmd
    Modified:   code/function/functions_fitting_Gaussian.R
    Modified:   code/function/functions_simulation.R
    Modified:   output/example/L_matrix.rda
    Modified:   output/example/fitted_datasets.rda
    Modified:   output/example/functional_datasets.rda
    Modified:   output/example/posterior_weights_matrix.rda
    Deleted:    output/simulation/L_matrix.rda
    Deleted:    output/simulation/fitted_datasets.rda
    Deleted:    output/simulation/posterior_weights_matrix.rda

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/simulation_pollution.Rmd</code>)
and HTML (<code>docs/simulation_pollution.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/8928abfb3ba7fdc1dfae44105e2134de600afab4/docs/simulation_pollution.html" target="_blank">8928abf</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/aa79b74e85d316d7c1491ad055dd44e4226127e0/analysis/simulation_pollution.Rmd" target="_blank">aa79b74</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
<td>
wflow_publish(files = "analysis/simulation_pollution.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/49dbb1251a432a05925eb234d78092703335c669/docs/simulation_pollution.html" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/d24fc30fe35a695ed101efcadb51ddbf32b2f99c/analysis/simulation_pollution.Rmd" target="_blank">d24fc30</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
<td>
wflow_publish(files = "analysis/simulation_pollution.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="setup" class="section level3">
<h3><strong>Setup</strong></h3>
<p>In this simulation, we consider the following model: <span
class="math display">\[y_i(x_j) \sim \mathcal{N}\big(f_i(x_j),
0.1\big),\]</span> where <span class="math inline">\(y_i(x_j)\)</span>
denotes the measured concentration of the <span
class="math inline">\(i\)</span>-th pollutant at the <span
class="math inline">\(j\)</span>-th time point, and <span
class="math inline">\(f_i\)</span> denotes the true concentration of
that pollutant. The measurement is assumed to have a known standard
deviation of <span class="math inline">\(0.1\)</span>.</p>
<p>We observe <span class="math inline">\(N = 48\)</span> pollutants at
<span class="math inline">\(n = 100\)</span> time points. The time
points <span class="math inline">\(\boldsymbol{x}\)</span> for each
pollutant are generated from a uniform distribution on the interval
<span class="math inline">\([0, 5]\)</span>.</p>
<p>We are interested in inferring the true concentration of each
pollutant as well as grouping them into similar categories.</p>
<p>We assign each true concentration function an independent prior,
which is a finite mixture of second order IWP with different <span
class="math inline">\(\sigma_k\)</span>. So <span
class="math display">\[f_i |\pi_0,..,\pi_K \overset{d}{=} \sum_{k=0}^K
\pi_k \text{GP}_k,\]</span> where each mixture component <span
class="math inline">\(\text{GP}_k\)</span> is defined as <span
class="math display">\[\frac{\partial^2}{\partial x^2}f(x) = \sigma_k
W(x).\]</span></p>
</div>
<div id="data" class="section level3">
<h3><strong>Data</strong></h3>
<p>We assume there are three true categories of pollutants (A,B and C),
and the true concentration of each pollutant is generated from a random
linear combination of the cubic-B spline basis.</p>
<ul>
<li>In category A, the true concentration is generated from the cubic-B
spline basis with 5 knots.</li>
<li>In category B, the true concentration is generated from the cubic-B
spline basis with 10 knots.</li>
<li>In category C, the true concentration is generated from the cubic-B
spline basis with 20 knots.</li>
</ul>
<pre class="r"><code># Generate indices of groups
group_indices &lt;- rep(1:3, each = 16)

# Generate datasets in parallel
datasets &lt;- mclapply(1:48, function(i) {
  set.seed(i)
  if (i &lt;= 16) {
    n_basis &lt;- 5
    sd_fun &lt;- 1
  } else if (i &lt;= 32) {
    n_basis &lt;- 10
    sd_fun &lt;- 1
  } else {
    n_basis &lt;- 20
    sd_fun &lt;- 1
  }
  simulate_process(n = 100, n_basis = n_basis, sd_fun = sd_fun, sd = 0.1)
}, mc.cores = num_cores)</code></pre>
<p>The observed time series for each category are shown below
(observations in points, true concentration functions in red).</p>
<p>Category A:</p>
<pre class="r"><code>par(mfrow = c(4, 4), mar = c(2, 2, 1, 1))  
for (i in 1:16) {
  plot(datasets[[i]]$x, datasets[[i]]$y, type = &#39;p&#39;,  
       main = paste(&quot;Dataset&quot;, i), xlab = &quot;x&quot;, ylab = &quot;y&quot;,
       cex = 0.1,
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)  
  lines(datasets[[i]]$x, datasets[[i]]$truef, col = &quot;red&quot;)
}</code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-3-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Category B:</p>
<pre class="r"><code>par(mfrow = c(4, 4), mar = c(2, 2, 1, 1))  
for (i in 17:32) {
  plot(datasets[[i]]$x, datasets[[i]]$y, type = &#39;p&#39;,  
       main = paste(&quot;Dataset&quot;, i), xlab = &quot;x&quot;, ylab = &quot;y&quot;,
       cex = 0.1,
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)  
  lines(datasets[[i]]$x, datasets[[i]]$truef, col = &quot;red&quot;)
}</code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-4-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Category C:</p>
<pre class="r"><code>par(mfrow = c(4, 4), mar = c(2, 2, 1, 1))  
for (i in 33:48) {
  plot(datasets[[i]]$x, datasets[[i]]$y, type = &#39;p&#39;,  
       main = paste(&quot;Dataset&quot;, i), xlab = &quot;x&quot;, ylab = &quot;y&quot;,
       cex = 0.1,
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)  
  lines(datasets[[i]]$x, datasets[[i]]$truef, col = &quot;red&quot;)
}</code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-5-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="empirical-bayes" class="section level3">
<h3><strong>Empirical Bayes</strong></h3>
<p>We setup the grid of <span class="math inline">\(\sigma_k\)</span> as
the following: (actually, a scaled version of it, called the predictive
standard deviation (PSD)).</p>
<pre class="r"><code>set.seed(123)
p_vec &lt;- 2
log_prec &lt;- unique(sort(c(Inf, seq(-1,1, by = 0.1), seq(-5,-1, by = 0.5), seq(1,5, by = 0.5), seq(-10,-5, by = 1), seq(5,10, by = 1)), decreasing = T))
psd_iwp_vec &lt;- 1/exp(.5*log_prec)

# Using the FEM:
L_vecs &lt;- mclapply(datasets, function(dataset) {
  compute_log_likelihood_ospline_seq2(
    x = dataset$x,
    y = dataset$y,
    p = p_vec,
    num_knots = 50,
    psd_iwp_vector = psd_iwp_vec,
    pred_step = 1,
    betaprec = 0.001,
    sd_gaussian = 0.1
  )
}, mc.cores = num_cores)
L_matrix &lt;- do.call(rbind, L_vecs)


# ## Using the exact method:
# L_vecs &lt;- mclapply(datasets, function(dataset) {
#   compute_log_likelihood_exact_aug_seq(
#     x = dataset$x,
#     y = dataset$y,
#     p = p_vec,
#     psd_iwp_vector = psd_iwp_vec,
#     pred_step = 1,
#     betaprec = 0.001,
#     sd_gaussian = 0.1
#   )
# }, mc.cores = num_cores)
# L_matrix &lt;- do.call(rbind, L_vecs)

save(L_matrix, file = paste0(result_dir, &quot;/L_matrix.rda&quot;))</code></pre>
<p>Then, we make use of <span
class="math inline">\(\texttt{mixsqp}\)</span> to estimate the prior
weights:</p>
<pre class="r"><code># which row of L has nan
nan_rows &lt;- which(apply(L_matrix, 1, function(x) any(is.nan(x))))
if(length(nan_rows) &gt; 0) {
  ## warn how many rows being removed
  warning(paste(&quot;Removing&quot;, length(nan_rows), &quot;rows with NaN&quot;))
  L_matrix &lt;- L_matrix[-nan_rows,]
}

fit.sqp &lt;- mixsqp(L = L_matrix, log = TRUE)</code></pre>
<pre><code>Running mix-SQP algorithm 0.3-54 on 48 x 48 matrix
convergence tol. (SQP):     1.0e-08
conv. tol. (active-set):    1.0e-10
zero threshold (solution):  1.0e-08
zero thresh. (search dir.): 1.0e-14
l.s. sufficient decrease:   1.0e-02
step size reduction factor: 7.5e-01
minimum step size:          1.0e-08
max. iter (SQP):            1000
max. iter (active-set):     20
number of EM iterations:    10
Computing SVD of 48 x 48 matrix.
Matrix is not low-rank; falling back to full matrix.
iter        objective max(rdual) nnz stepsize max.diff nqp nls
   1 +2.059242743e+00  -- EM --   43 1.00e+00 9.37e-02  --  --
   2 +2.005627596e+00  -- EM --   43 1.00e+00 2.53e-02  --  --
   3 +1.985920171e+00  -- EM --   39 1.00e+00 1.39e-02  --  --
   4 +1.976259235e+00  -- EM --   39 1.00e+00 1.08e-02  --  --
   5 +1.970650031e+00  -- EM --   38 1.00e+00 8.89e-03  --  --
   6 +1.966996657e+00  -- EM --   38 1.00e+00 7.66e-03  --  --
   7 +1.964413615e+00  -- EM --   37 1.00e+00 6.80e-03  --  --
   8 +1.962472449e+00  -- EM --   37 1.00e+00 6.18e-03  --  --
   9 +1.960944866e+00  -- EM --   37 1.00e+00 5.69e-03  --  --
  10 +1.959700250e+00  -- EM --   36 1.00e+00 5.30e-03  --  --
   1 +1.958659440e+00 +3.036e-02  35  ------   ------   --  --
   2 +1.952467076e+00 +1.306e-02  15 1.00e+00 9.93e-02  20   1
   3 +1.949525509e+00 +1.220e-03  11 1.00e+00 1.05e-01   9   1
   4 +1.949524527e+00 +1.024e-06  11 1.00e+00 1.96e-03   2   1
   5 +1.949524527e+00 -5.829e-08  11 1.00e+00 2.76e-06   2   1
Optimization took 0.00 seconds.
Convergence criteria met---optimal solution found.</code></pre>
<pre class="r"><code>numiter &lt;- nrow(fit.sqp$progress)
plot(1:numiter,fit.sqp$progress$objective,type = &quot;b&quot;,
     pch = 20,lwd = 2,xlab = &quot;SQP iteration&quot;,
     ylab = &quot;objective&quot;,xaxp = c(1,numiter,numiter - 1))</code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-8-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>prior_weight &lt;- data.frame(p = rep(p_vec, each = length(psd_iwp_vec)), psd_iwp = psd_iwp_vec, prior_weight = fit.sqp$x)</code></pre>
<p>We can take a look at the estimated prior:</p>
<pre class="r"><code>head(prior_weight)</code></pre>
<pre><code>  p     psd_iwp prior_weight
1 2 0.000000000   0.02077435
2 2 0.006737947   0.00000000
3 2 0.011108997   0.00000000
4 2 0.018315639   0.00000000
5 2 0.030197383   0.00000000
6 2 0.049787068   0.00000000</code></pre>
</div>
<div id="posterior-inference" class="section level3">
<h3><strong>Posterior Inference</strong></h3>
<p>With the estimated prior, we can now perform the posterior inference
for each dataset:</p>
<pre class="r"><code>num_datasets &lt;- length(datasets)
num_weights &lt;- sum(prior_weight$prior_weight != 0)
posterior_weights_matrix &lt;- matrix(nrow = num_datasets, ncol = num_weights)

# Loop through each dataset and perform fitting
fitted_datasets &lt;- list()

# Using FEM:
for (i in seq_along(datasets)) {
  dataset &lt;- datasets[[i]]
  fit_result_final &lt;- fit_ospline_with_prior2(
    num_cores = num_cores,
    x = dataset$x,
    y = dataset$y,
    num_knots = 50,
    prior_weight = prior_weight,
    betaprec = 0.001,
    sd_gaussian = 0.1,
    pred_step = 1
  )
  fitted_datasets[[i]] &lt;- aggregate_fit_with_prior(x = dataset$x, fit_results_with_prior = fit_result_final)$summary_df
  posterior_weights_matrix[i, ] &lt;- fit_result_final$posterior_weights[, &quot;posterior_weight&quot;]
}

# # Using the exact method:
# for (i in seq_along(datasets)) {
#   dataset &lt;- datasets[[i]]
#   fit_result_final &lt;- fit_exact_aug_with_prior2(
#     num_cores = num_cores,
#     x = dataset$x,
#     y = dataset$y,
#     prior_weight = prior_weight,
#     betaprec = 0.001,
#     sd_gaussian = 0.1,
#     pred_step = 1
#   )
#   fitted_datasets[[i]] &lt;- aggregate_fit_with_prior(x = dataset$x, fit_results_with_prior = fit_result_final)$summary_df
#   posterior_weights_matrix[i, ] &lt;- fit_result_final$posterior_weights[, &quot;posterior_weight&quot;]
# }

colnames(posterior_weights_matrix) &lt;- paste(as.character(fit_result_final$posterior_weights[, &quot;p&quot;]),
                                            as.character(fit_result_final$posterior_weights[, &quot;psd_iwp&quot;]), sep = &quot;_&quot;)

save(posterior_weights_matrix, file = paste0(result_dir, &quot;/posterior_weights_matrix.rda&quot;))

save(fitted_datasets, file = paste0(result_dir, &quot;/fitted_datasets.rda&quot;))</code></pre>
<p>We can visualize the posterior weights for each dataset:</p>
<pre class="r"><code>posterior_weights_df &lt;- as.data.frame(posterior_weights_matrix)
posterior_weights_df$id &lt;- 1:nrow(posterior_weights_df)
melted_data &lt;- melt(posterior_weights_df, id.vars = &quot;id&quot;)
melted_data$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data$variable)
melted_data$variable3 &lt;- as.factor(round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data$variable)), 3))

ggplot(melted_data, aes(x = as.factor(id), y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +
  labs(x = &quot;Observation ID&quot;, y = &quot;Weight&quot;, fill = &quot;PSD&quot;) +
  theme_minimal() +
  ggtitle(&quot;Structure Plot of Posterior Weights&quot;) +
  coord_flip() </code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-1">
Past versions of unnamed-chunk-12-1.png
</button>
</p>
<div id="fig-unnamed-chunk-12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-12-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The fitted curves for pollutants in category A:</p>
<pre class="r"><code>par(mfrow = c(4, 4), mar = c(2, 2, 1, 1))  
for (i in 1:16) {
  plot(datasets[[i]]$x, datasets[[i]]$y, type = &#39;p&#39;,  
       main = paste(&quot;Dataset&quot;, i), xlab = &quot;x&quot;, ylab = &quot;y&quot;,
       cex = 0.2,
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)  
  lines(datasets[[i]]$x, datasets[[i]]$truef, col = &quot;red&quot;, lty = &quot;dashed&quot;, lwd = 1.5)
  lines(fitted_datasets[[i]]$x, fitted_datasets[[i]]$mean, col = &quot;blue&quot;, lwd = 1.2)
}</code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-13-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The fitted curves for pollutants in category B:</p>
<pre class="r"><code>par(mfrow = c(4, 4), mar = c(2, 2, 1, 1))  
for (i in 17:32) {
  plot(datasets[[i]]$x, datasets[[i]]$y, type = &#39;p&#39;,  
       main = paste(&quot;Dataset&quot;, i), xlab = &quot;x&quot;, ylab = &quot;y&quot;,
       cex = 0.2,
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)  
  lines(datasets[[i]]$x, datasets[[i]]$truef, col = &quot;red&quot;, lty = &quot;dashed&quot;, lwd = 1.5)
  lines(fitted_datasets[[i]]$x, fitted_datasets[[i]]$mean, col = &quot;blue&quot;, lwd = 1.2)
}</code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-14-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The fitted curves for pollutants in category C:</p>
<pre class="r"><code>par(mfrow = c(4, 4), mar = c(2, 2, 1, 1))  
for (i in 33:48) {
  plot(datasets[[i]]$x, datasets[[i]]$y, type = &#39;p&#39;,  
       main = paste(&quot;Dataset&quot;, i), xlab = &quot;x&quot;, ylab = &quot;y&quot;,
       cex = 0.2,
       cex.main = 0.8, cex.lab = 0.7, cex.axis = 0.7)  
  lines(datasets[[i]]$x, datasets[[i]]$truef, col = &quot;red&quot;, lty = &quot;dashed&quot;, lwd = 1.5)
  lines(fitted_datasets[[i]]$x, fitted_datasets[[i]]$mean, col = &quot;blue&quot;, lwd = 1.2)
}</code></pre>
<p><img src="figure/simulation_pollution.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-1">
Past versions of unnamed-chunk-15-1.png
</button>
</p>
<div id="fig-unnamed-chunk-15-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/49dbb1251a432a05925eb234d78092703335c669/docs/figure/simulation_pollution.Rmd/unnamed-chunk-15-1.png" target="_blank">49dbb12</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.0   dplyr_1.1.3    
 [5] purrr_1.0.2     readr_2.1.4     tidyr_1.3.0     tibble_3.2.1   
 [9] tidyverse_2.0.0 mixsqp_0.3-54   reshape2_1.4.4  ggplot2_3.5.1  
[13] Matrix_1.6-3    TMB_1.9.7       BayesGP_0.1.1  

loaded via a namespace (and not attached):
 [1] sass_0.4.7        utf8_1.2.3        generics_0.1.3    stringi_1.7.12   
 [5] lattice_0.21-8    hms_1.1.3         digest_0.6.33     magrittr_2.0.3   
 [9] timechange_0.2.0  evaluate_0.22     grid_4.3.1        fastmap_1.1.1    
[13] plyr_1.8.9        rprojroot_2.0.3   workflowr_1.7.1   jsonlite_1.8.7   
[17] whisker_0.4.1     promises_1.2.1    fansi_1.0.5       scales_1.3.0     
[21] jquerylib_0.1.4   cli_3.6.1         rlang_1.1.1       munsell_0.5.0    
[25] withr_2.5.1       cachem_1.0.8      yaml_2.3.7        tools_4.3.1      
[29] tzdb_0.4.0        colorspace_2.1-0  httpuv_1.6.11     vctrs_0.6.4      
[33] R6_2.5.1          lifecycle_1.0.3   git2r_0.33.0      fs_1.6.3         
[37] irlba_2.3.5.1     pkgconfig_2.0.3   pillar_1.9.0      bslib_0.5.1      
[41] later_1.3.1       gtable_0.3.4      glue_1.6.2        Rcpp_1.0.11      
[45] xfun_0.40         tidyselect_1.2.0  rstudioapi_0.15.0 knitr_1.44       
[49] farver_2.1.1      htmltools_0.5.6.1 labeling_0.4.3    rmarkdown_2.25   
[53] compiler_4.3.1   </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
