<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ziang Zhang" />

<meta name="date" content="2024-05-19" />

<title>Identifying dynamic eQTLs using FASH</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">resultsummary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="simulation.html">Simulation 1</a>
</li>
<li>
  <a href="simulation_eQTL_identification.html">Simulation 2</a>
</li>
<li>
  <a href="covid_example.html">COVID-19 Example</a>
</li>
<li>
  <a href="expression.html">Gene Expression Example</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/AgueroZZ/resultsummary">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Identifying dynamic eQTLs using FASH</h1>
<h4 class="author">Ziang Zhang</h4>
<h4 class="date">2024-05-19</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-05-24
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>FASHresultsummary/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240507code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240507)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240507code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240507)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreea53277262c0b48ae3703879a49f9f3a5c5fd6e4ftargetblanka532772a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f" target="_blank">a532772</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreea53277262c0b48ae3703879a49f9f3a5c5fd6e4ftargetblanka532772a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f" target="_blank">a532772</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    code/.DS_Store
    Ignored:    code/cpp/.DS_Store
    Ignored:    code/function/.DS_Store
    Ignored:    code/function/.Rhistory
    Ignored:    data/.DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/example/.DS_Store
    Ignored:    output/example/figure/
    Ignored:    output/simulation/.DS_Store
    Ignored:    output/simulation/figure/

Untracked files:
    Untracked:  code/cpp/Gaussian_just_fixed.o
    Untracked:  code/cpp/Gaussian_just_fixed.so
    Untracked:  code/cpp/Gaussian_theta_known.o
    Untracked:  code/cpp/Gaussian_theta_known.so
    Untracked:  code/cpp/Poisson.o
    Untracked:  code/cpp/Poisson.so
    Untracked:  code/cpp/Poisson_causes.o
    Untracked:  code/cpp/Poisson_causes.so
    Untracked:  code/cpp/Poisson_covid.o
    Untracked:  code/cpp/Poisson_covid.so
    Untracked:  code/cpp/Poisson_just_fixed.cpp
    Untracked:  code/cpp/Poisson_just_fixed.o
    Untracked:  code/cpp/Poisson_just_fixed.so
    Untracked:  code/cpp/Poisson_just_fixed_causes.o
    Untracked:  code/cpp/Poisson_just_fixed_causes.so
    Untracked:  code/cpp/Poisson_just_fixed_covid.o
    Untracked:  code/cpp/Poisson_just_fixed_covid.so
    Untracked:  code/function/functions_fitting_Gaussian_eQTL_irregular.R
    Untracked:  code/function/functions_simulation_eQTL_irregular.R
    Untracked:  data/13100810.csv
    Untracked:  data/13100810_MetaData.csv
    Untracked:  data/X13100810.eng.zip
    Untracked:  data/expression_data/
    Untracked:  output/causes/
    Untracked:  output/expression/
    Untracked:  output/simulation_eQTL/
    Untracked:  output/simulation_eQTL_test/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/simulation_eQTL_identification.Rmd</code>) and HTML
(<code>docs/simulation_eQTL_identification.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/analysis/simulation_eQTL_identification.Rmd" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
<td>
update eqtl sim
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/simulation_eQTL_identification.html" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
<td>
update eqtl sim
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/9e1131d55208c45512bc391146047aa3a355c8f8/docs/simulation_eQTL_identification.html" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/513eb20c5a9c05196b72f000ddb0ec117aff006f/analysis/simulation_eQTL_identification.Rmd" target="_blank">513eb20</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
<td>
wflow_git_commit(c("code/function/functions_fitting_Gaussian_eQTL_test.R",
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="context" class="section level2">
<h2><strong>Context</strong></h2>
<p>We consider the effect size estimate of <span
class="math inline">\(1000\)</span> eQTLs measured in day <span
class="math inline">\(t = 1\)</span> to day <span
class="math inline">\(t = 16\)</span>:</p>
<ul>
<li>there are <span class="math inline">\(700\)</span> eQTLs that are
not dynamic, i.e., the effect size is constant over time.</li>
<li>there are <span class="math inline">\(200\)</span> eQTLs that are
linear dynamic, i.e., the effect size is changing linearly over
time.</li>
<li>there are <span class="math inline">\(100\)</span> eQTLs that are
non-linear dynamic, i.e., the effect size is changing non-linearly over
time.</li>
</ul>
<p>For simplicity, let’s assume the SE of effect estimate is constant
over time: <span class="math inline">\(\text{SE}(\hat{\beta}_{i}(t_j)) =
\sigma_i, \forall j\)</span> and hence: <span
class="math display">\[\hat{\beta}_i(t_j) \sim
N(\beta_i(t_j),\sigma^2).\]</span></p>
<p>We consider the following prior: <span
class="math display">\[\beta_i(t) = \sum_{j=0}^K \pi_j g_j(t),\]</span>
where each <span class="math inline">\(g_j(t)\)</span> is a GP defined
by: <span class="math display">\[Lg_j(t) = \sigma_j W(t),\]</span> with
<span class="math inline">\(L = \frac{d}{dt}\)</span> and <span
class="math inline">\(W(t)\)</span> denotes the Gaussian white noise. In
this case, the base model is the space of constant functions <span
class="math inline">\(\text{span}\{1\}\)</span>. The size of <span
class="math inline">\(\sigma_j\)</span> controls the strength of the
shrinkage towards the base model. The first component has <span
class="math inline">\(\sigma_0 = 0\)</span> to represent the base
model.</p>
</div>
<div id="data" class="section level2">
<h2><strong>Data</strong></h2>
<p>Load the required libraries and set the seed.</p>
<pre class="r"><code>library(BayesGP)
library(TMB)
library(Matrix)
library(splines)
library(parallel)
library(ggplot2)
library(reshape2)
library(mixsqp)
library(tidyverse)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ lubridate 1.9.3     ✔ tibble    3.2.1
✔ purrr     1.0.2     ✔ tidyr     1.3.0
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ tidyr::expand() masks Matrix::expand()
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
✖ tidyr::pack()   masks Matrix::pack()
✖ tidyr::unpack() masks Matrix::unpack()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<pre class="r"><code>library(mashr)</code></pre>
<pre><code>Loading required package: ashr</code></pre>
<pre class="r"><code>cpp_dir &lt;- paste0(getwd(), &quot;/code/cpp&quot;)
fig_dir &lt;- paste0(getwd(), &quot;/output/simulation_eQTL_test&quot;)
result_dir &lt;- paste0(getwd(), &quot;/output/simulation_eQTL_test&quot;)
function_dir &lt;- paste0(getwd(), &quot;/code/function&quot;)
source(paste0(function_dir, &quot;/functions_fitting_Gaussian_eQTL_test.R&quot;))
source(paste0(function_dir, &quot;/functions_simulation_eQTL_test.R&quot;))
compile(paste0(cpp_dir, &quot;/Gaussian_theta_known.cpp&quot;))</code></pre>
<pre><code>Warning in readLines(file): incomplete final line found on
&#39;/Users/ziangzhang/Desktop/FASH/FASHresultsummary/code/cpp/Gaussian_theta_known.cpp&#39;</code></pre>
<pre><code>Note: Using Makevars in /Users/ziangzhang/.R/Makevars </code></pre>
<pre><code>using C++ compiler: &#39;Apple clang version 13.1.6 (clang-1316.0.21.2.5)&#39;
using SDK: &#39;MacOSX12.3.sdk&#39;</code></pre>
<pre><code>[1] 0</code></pre>
<pre class="r"><code>compile(paste0(cpp_dir, &quot;/Gaussian_just_fixed.cpp&quot;))</code></pre>
<pre><code>Warning in readLines(file): incomplete final line found on
&#39;/Users/ziangzhang/Desktop/FASH/FASHresultsummary/code/cpp/Gaussian_just_fixed.cpp&#39;</code></pre>
<pre><code>Note: Using Makevars in /Users/ziangzhang/.R/Makevars </code></pre>
<pre><code>using C++ compiler: &#39;Apple clang version 13.1.6 (clang-1316.0.21.2.5)&#39;
using SDK: &#39;MacOSX12.3.sdk&#39;</code></pre>
<pre><code>[1] 0</code></pre>
<pre class="r"><code>dyn.load(TMB::dynlib(paste0(cpp_dir, &quot;/Gaussian_theta_known&quot;)))
dyn.load(TMB::dynlib(paste0(cpp_dir, &quot;/Gaussian_just_fixed&quot;)))</code></pre>
<pre class="r"><code>num_cores &lt;- 4
set.seed(123)
N &lt;- 1000
sigma &lt;- sample(c(0.1, 0.5, 1), size = N, replace = TRUE, prob = c(1/3,1/3,1/3))
set.seed(123)</code></pre>
<p>First, we simulate <span class="math inline">\(700\)</span>
non-dynamic eQTLs:</p>
<pre class="r"><code>sizeA &lt;- 700
data_sim_list_A &lt;- lapply(1:sizeA, function(i) simulate_process(sd_poly = 1, type = &quot;nondynamic&quot;, sd = sigma[i]))
par(mfrow = c(2, 2))
for (i in 1:4) {
  plot(data_sim_list_A[[i]]$x, data_sim_list_A[[i]]$truef, 
       type = &quot;l&quot;, col = &quot;red&quot;, xlab = &quot;Time&quot;,
       ylim = range(data_sim_list_A[[i]]$truef, data_sim_list_A[[i]]$y),
       ylab = &quot;Effect size&quot;, main = paste0(&quot;eQTL &quot;, i))
  points(data_sim_list_A[[i]]$x, data_sim_list_A[[i]]$y, col = &quot;blue&quot;)
}</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-3-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>par(mfrow = c(1, 1))</code></pre>
<p>Then, we simulate <span class="math inline">\(200\)</span> dynamic
eQTLs with linear dynamics:</p>
<pre class="r"><code>sizeB &lt;- 200
data_sim_list_B &lt;- lapply(1:sizeB, function(i) simulate_process(sd_poly = 0.1, type = &quot;linear&quot;, sd = sigma[i + sizeA]))
par(mfrow = c(2, 2))
for (i in 1:4) {
  plot(data_sim_list_B[[i]]$x, data_sim_list_B[[i]]$truef, 
       type = &quot;l&quot;, col = &quot;red&quot;, xlab = &quot;Time&quot;,
       ylim = range(data_sim_list_B[[i]]$truef, data_sim_list_B[[i]]$y),
       ylab = &quot;Effect size&quot;, main = paste0(&quot;eQTL &quot;, i))
  points(data_sim_list_B[[i]]$x, data_sim_list_B[[i]]$y, col = &quot;blue&quot;)
}</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-4-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>par(mfrow = c(1, 1))</code></pre>
<p>Finally, simulate <span class="math inline">\(100\)</span> non-linear
dynamic eQTLs.</p>
<pre class="r"><code>sizeC &lt;- 100
data_sim_list_C &lt;- lapply(1:sizeC, function(i) simulate_process(sd_poly = 0.1, type = &quot;nonlinear&quot;, sd = sigma[i + sizeA + sizeB], sd_fun = 1))
par(mfrow = c(2, 2))
for (i in 1:4) {
  plot(data_sim_list_C[[i]]$x, data_sim_list_C[[i]]$truef, 
       type = &quot;l&quot;, col = &quot;red&quot;, xlab = &quot;Time&quot;,
       ylim = range(data_sim_list_C[[i]]$truef, data_sim_list_C[[i]]$y),
       ylab = &quot;Effect size&quot;, main = paste0(&quot;eQTL &quot;, i))
  points(data_sim_list_C[[i]]$x, data_sim_list_C[[i]]$y, col = &quot;blue&quot;)
}</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-5-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>par(mfrow = c(1, 1))

datasets &lt;- c(data_sim_list_A, data_sim_list_B, data_sim_list_C)
sigma &lt;- unlist(lapply(datasets, function(x) unique(x$sd)))</code></pre>
</div>
<div id="hypothesis-testing" class="section level2">
<h2><strong>Hypothesis Testing</strong></h2>
<div id="which-eqtls-are-non-dynamic" class="section level3">
<h3><strong>Which eQTLs are non-dynamic?</strong></h3>
<p>Let’s consider the question of identifying dynamic eQTLs. Based on
the prior we used, an eQTL should be considered non-dynamic if the
posterior weight <span class="math inline">\(\pi_0\)</span> of the first
component is close to <span class="math inline">\(1\)</span>.</p>
<p>First, we compute the L matrix:</p>
<pre class="r"><code>set.seed(123)
p_vec &lt;- 1
psd_iwp_vec &lt;- sort(unique(c(0,seq(0,1, by = 0.05))))
L_vecs &lt;- list()
# create a progress bar
pb &lt;- txtProgressBar(min = 0, max = length(datasets), style = 3)
for (i in 1:length(datasets)) {
  setTxtProgressBar(pb, i)
  L_vecs[[i]] &lt;- compute_log_likelihood_ospline_seq2(
    x = datasets[[i]]$x,
    y = datasets[[i]]$y,
    p = p_vec,
    num_knots = 16,
    psd_iwp_vector = psd_iwp_vec,
    pred_step = 1,
    betaprec = 0.001,
    sd_gaussian = sigma[i]
  )
}
L_matrix &lt;- do.call(rbind, L_vecs)
save(L_matrix, file = paste0(result_dir, &quot;/L_matrix.rda&quot;))</code></pre>
<p>Based on the L-matrix, we optimize the prior weights through EB:</p>
<pre class="r"><code>fit.sqp &lt;- mixsqp(L = L_matrix, log = TRUE)</code></pre>
<pre><code>Running mix-SQP algorithm 0.3-54 on 1000 x 21 matrix
convergence tol. (SQP):     1.0e-08
conv. tol. (active-set):    1.0e-10
zero threshold (solution):  1.0e-08
zero thresh. (search dir.): 1.0e-14
l.s. sufficient decrease:   1.0e-02
step size reduction factor: 7.5e-01
minimum step size:          1.0e-08
max. iter (SQP):            1000
max. iter (active-set):     20
number of EM iterations:    10
Computing SVD of 1000 x 21 matrix.
Matrix is not low-rank; falling back to full matrix.
iter        objective max(rdual) nnz stepsize max.diff nqp nls
   1 +6.383868942e-01  -- EM --   21 1.00e+00 6.56e-01  --  --
   2 +5.501865119e-01  -- EM --   21 1.00e+00 1.55e-01  --  --
   3 +5.454211990e-01  -- EM --   21 1.00e+00 1.16e-02  --  --
   4 +5.442450364e-01  -- EM --   21 1.00e+00 1.49e-03  --  --
   5 +5.438115214e-01  -- EM --   21 1.00e+00 8.59e-04  --  --
   6 +5.436096886e-01  -- EM --   21 1.00e+00 4.96e-04  --  --
   7 +5.434969163e-01  -- EM --   21 1.00e+00 3.05e-04  --  --
   8 +5.434241366e-01  -- EM --   21 1.00e+00 2.02e-04  --  --
   9 +5.433718212e-01  -- EM --   21 1.00e+00 1.60e-04  --  --
  10 +5.433312684e-01  -- EM --   21 1.00e+00 1.51e-04  --  --
   1 +5.432981976e-01 +3.246e-02  21  ------   ------   --  --
   2 +5.427690931e-01 +4.447e-03   7 1.00e+00 1.50e-02  20   1
   3 +5.427557262e-01 +3.102e-04   8 1.00e+00 7.82e-03   4   1
   4 +5.427557083e-01 +2.040e-07   8 1.00e+00 2.87e-04   2   1
   5 +5.427557083e-01 -1.649e-08   8 1.00e+00 4.67e-07   2   1
Optimization took 0.00 seconds.
Convergence criteria met---optimal solution found.</code></pre>
<pre class="r"><code>numiter &lt;- nrow(fit.sqp$progress)
plot(1:numiter,fit.sqp$progress$objective,type = &quot;b&quot;,
     pch = 20,lwd = 2,xlab = &quot;SQP iteration&quot;,
     ylab = &quot;objective&quot;,xaxp = c(1,numiter,numiter - 1))</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-8-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>prior_weight &lt;- data.frame(p = rep(p_vec, each = length(psd_iwp_vec)), psd_iwp = psd_iwp_vec, prior_weight = fit.sqp$x)</code></pre>
<pre class="r"><code>head(prior_weight)</code></pre>
<pre><code>  p psd_iwp prior_weight
1 1    0.00   0.87107064
2 1    0.05   0.00000000
3 1    0.10   0.01652265
4 1    0.15   0.03652852
5 1    0.20   0.02047323
6 1    0.25   0.02995007</code></pre>
<p>The estimated overall proportion of non-dynamic eQTLs is:</p>
<pre class="r"><code>prior_weight %&gt;%
  filter(psd_iwp == 0) %&gt;%
  pull(prior_weight)</code></pre>
<pre><code>[1] 0.8710706</code></pre>
<p>With the estimated prior, we can now perform the posterior inference
for each dataset:</p>
<pre class="r"><code>num_datasets &lt;- length(datasets)
num_weights &lt;- sum(prior_weight$prior_weight != 0)
posterior_weights_matrix &lt;- matrix(nrow = num_datasets, ncol = num_weights)

# Loop through each dataset and perform fitting
fitted_datasets &lt;- list()
# start a progress bar
pb &lt;- txtProgressBar(min = 0, max = num_datasets, style = 3)
for (i in seq_along(datasets)) {
  setTxtProgressBar(pb, i)
  dataset &lt;- datasets[[i]]
  fit_result_final &lt;- fit_ospline_with_prior2(
    num_cores = 1,
    x = dataset$x,
    y = dataset$y,
    num_knots = 16,
    prior_weight = prior_weight,
    betaprec = 0.001,
    sd_gaussian = sigma[i],
    pred_step = 1
  )
  fitted_datasets[[i]] &lt;- aggregate_fit_with_prior(x = dataset$x, fit_results_with_prior = fit_result_final)$summary_df
  posterior_weights_matrix[i, ] &lt;- fit_result_final$posterior_weights[, &quot;posterior_weight&quot;]
}
colnames(posterior_weights_matrix) &lt;- paste(as.character(fit_result_final$posterior_weights[, &quot;p&quot;]),
                                            as.character(fit_result_final$posterior_weights[, &quot;psd_iwp&quot;]), sep = &quot;_&quot;)
save(posterior_weights_matrix, file = paste0(result_dir, &quot;/posterior_weights_matrix.rda&quot;))
save(fitted_datasets, file = paste0(result_dir, &quot;/fitted_datasets.rda&quot;))</code></pre>
<p>We can visualize the posterior weights for each dataset:</p>
<pre class="r"><code>posterior_weights_df &lt;- as.data.frame(posterior_weights_matrix)
posterior_weights_df$id &lt;- c(1:length(sigma))
melted_data &lt;- melt(posterior_weights_df, id.vars = &quot;id&quot;)
melted_data$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data$variable)
melted_data$variable3 &lt;- (round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data$variable)), 3))

ggplot(melted_data, aes(x = id, y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +
  labs(x = &quot;Observation ID&quot;, y = &quot;Weight&quot;, fill = &quot;PSD&quot;) +
  theme_minimal() +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  ggtitle(&quot;FASH: Structure Plot of Posterior Weights&quot;) +
  coord_flip() </code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-13-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="using-fash" class="section level4">
<h4><em>Using FASH</em></h4>
<p>Compute the local false discovery rate (lfda):</p>
<pre class="r"><code>set.seed(123)
lfda &lt;- posterior_weights_matrix[,1]
fda_df &lt;- data.frame(eQTL = 1:length(lfda), fda = lfda, type = rep(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), times = c(sizeA, sizeB, sizeC)))
fda_df &lt;- fda_df[order(fda_df$fda), ] # ordering it
fda_df$cumulative_fda &lt;- cumsum(fda_df$fda)/seq_along(fda_df$fda)
fda_df$rank &lt;- 1:length(lfda)</code></pre>
<pre class="r"><code>ggplot(fda_df, aes(x = 1:length(lfda), y = cumulative_fda, col = type)) +
  geom_point() +
  geom_hline(yintercept = 0.05, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Ordered eQTLs&quot;, y = &quot;Cumulative FDA&quot;, col = &quot;Type&quot;) +
  theme_minimal() +
  ggtitle(&quot;Cumulative FDA Plot&quot;) +
  scale_color_manual(values = c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;))</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-1">
Past versions of unnamed-chunk-15-1.png
</button>
</p>
<div id="fig-unnamed-chunk-15-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-15-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-15-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>How many false discoveries are there?</p>
<pre class="r"><code>alpha &lt;- 0.05
num_discoveries &lt;- sum(fda_df$cumulative_fda &lt;= alpha)
num_false_discoveries &lt;- sum(fda_df$cumulative_fda &lt;= alpha &amp; fda_df$type == &quot;A&quot;)
true_false_discovery_rate &lt;- num_false_discoveries/num_discoveries
true_false_discovery_rate</code></pre>
<pre><code>[1] 0</code></pre>
<p>Plot the curve of nominal false discovery rate (threshold) against
the actual false discovery rate:</p>
<pre class="r"><code># Calculate true FDR for FASH
threshold_vec &lt;- seq(0.01, 0.99, by = 0.01)
fdr_vec &lt;- numeric(length(threshold_vec))

for (i in 1:length(threshold_vec)) {
  num_discoveries &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df$type == &quot;A&quot;)
  fdr_vec[i] &lt;- num_false_discoveries / num_discoveries
}

# Create a data frame for plotting
fdr_df_fash &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec)

# Plot the nominal FDR vs true FDR for FASH
ggplot(fdr_df_fash, aes(x = threshold, y = true_fdr)) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;(Nominal) False Discovery Rate&quot;, y = &quot;(Actual) False Discovery Rate&quot;) +
  theme_minimal() +
  geom_hline(yintercept = sizeA/N, linetype = &quot;dashed&quot;, color = &quot;red&quot;) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1)) +
  ggtitle(&quot;Nominal FDR vs Actual FDR Curve for FASH&quot;)</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-17-1">
Past versions of unnamed-chunk-17-1.png
</button>
</p>
<div id="fig-unnamed-chunk-17-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-17-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-17-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Also compare the number of false discoveries with the number of
discoveries at different level of threshold, where number of discoveries
is plotted against number of false discoveries:</p>
<pre class="r"><code>threshold_vec &lt;- seq(0.01, 0.99, by = 0.01)
num_discoveries_vec &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec[i] &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries_vec[i] &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df$type == &quot;A&quot;)
}
num_discoveries_df &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec, num_false_discoveries = num_false_discoveries_vec)
ggplot(num_discoveries_df, aes(x = (num_discoveries), y = (num_false_discoveries))) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Number of Discoveries&quot;, y = &quot;Number of False Discoveries&quot;) +
  theme_minimal() +
  ggtitle(&quot;FASH: Number of False Discoveries vs Number of Discoveries&quot;) +
  coord_cartesian(xlim = c(0, 300), ylim = c(0, 300))</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-18-1">
Past versions of unnamed-chunk-18-1.png
</button>
</p>
<div id="fig-unnamed-chunk-18-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-18-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-18-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="using-mash" class="section level4">
<h4><em>Using MASH</em></h4>
<p>Now, let’s do the comparison with the default implementation of
MASH:</p>
<pre class="r"><code>fitted_datasets_mash &lt;- list()

# Produce a huge data-matrix, the i-th row being dataset[[i]]$y
all_data_matrix &lt;- do.call(rbind, lapply(datasets, function(x) x$y))
SE_matrix &lt;- matrix(nrow = nrow(all_data_matrix), ncol = ncol(all_data_matrix), sigma)

# now use mashr:
mash_data &lt;-  mashr::mash_set_data(all_data_matrix, SE_matrix)
m.1by1 = mashr::mash_1by1(mash_data)
strong = mashr::get_significant_results(m.1by1, 0.05)
# keep the top 10%
strong &lt;- strong[1:round(0.1*length(strong))]
U.pca = mashr::cov_pca(mash_data, 5, subset = strong)
U.ed = cov_ed(mash_data, U.pca, subset=strong)
U.c = cov_canonical(mash_data)  
m   = mash(mash_data, c(U.c,U.ed))</code></pre>
<pre><code> - Computing 1000 x 595 likelihood matrix.
 - Likelihood calculations took 1.12 seconds.
 - Fitting model with 595 mixture components.
 - Model fitting took 8.13 seconds.
 - Computing posterior matrices.
 - Computation allocated took 0.09 seconds.</code></pre>
<pre class="r"><code>mash_post &lt;- m$posterior_weights
## extract the colnames start with &quot;equal_effects&quot;
lfda_mash &lt;- mash_post[, c(1, grep(&quot;equal_effects&quot;, colnames(mash_post)))]
# sum each row
lfda_mash &lt;- rowSums(lfda_mash)

fda_df_mash &lt;- data.frame(eQTL = 1:length(lfda), fda = lfda_mash, type = rep(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), times = c(sizeA, sizeB, sizeC)))

fda_df_mash &lt;- fda_df_mash[order(fda_df_mash$fda), ] # ordering it
fda_df_mash$cumulative_fda &lt;- cumsum(fda_df_mash$fda)/seq_along(fda_df_mash$fda)
fda_df_mash$rank &lt;- 1:length(lfda_mash)</code></pre>
<p>How many false discoveries are there?</p>
<pre class="r"><code>alpha &lt;- 0.05
num_discoveries &lt;- sum(fda_df_mash$cumulative_fda &lt;= alpha)
num_false_discoveries &lt;- sum(fda_df_mash$cumulative_fda &lt;= alpha &amp; fda_df_mash$type == &quot;A&quot;)
true_false_discovery_rate &lt;- num_false_discoveries/num_discoveries
true_false_discovery_rate</code></pre>
<pre><code>[1] 0.04324324</code></pre>
<pre class="r"><code>ggplot(fda_df_mash, aes(x = rank, y = cumulative_fda, col = type)) +
  geom_point() +
  geom_hline(yintercept = 0.05, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Ordered eQTLs&quot;, y = &quot;Cumulative FDA&quot;, col = &quot;Type&quot;) +
  theme_minimal() +
  coord_cartesian(ylim = c(0, 1)) +
  ggtitle(&quot;Cumulative FDA Plot&quot;) +
  scale_color_manual(values = c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;))</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-22-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code># Calculate true FDR for MASH
threshold_vec &lt;- seq(0.01, 0.99, by = 0.01)
fdr_vec_mash &lt;- numeric(length(threshold_vec))

for (i in 1:length(threshold_vec)) {
  num_discoveries_mash &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries_mash &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df_mash$type == &quot;A&quot;)
  fdr_vec_mash[i] &lt;- num_false_discoveries_mash / num_discoveries_mash
}

# Create a data frame for plotting
fdr_df_mash &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_mash)

# Plot the nominal FDR vs true FDR for MASH
ggplot(fdr_df_mash, aes(x = threshold, y = true_fdr)) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;(Nominal) False Discovery Rate&quot;, y = &quot;(Actual) False Discovery Rate&quot;) +
  theme_minimal() +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1)) +
  ggtitle(&quot;Nominal FDR vs Actual FDR Curve for MASH&quot;)</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-23-1">
Past versions of unnamed-chunk-23-1.png
</button>
</p>
<div id="fig-unnamed-chunk-23-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-23-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-23-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot the number of false discoveries against the number of total
discoveries for MASH:</p>
<pre class="r"><code>threshold_vec &lt;- seq(0.01, 0.99, by = 0.01)
num_discoveries_vec &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec[i] &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries_vec[i] &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df_mash$type == &quot;A&quot;)
}
num_discoveries_df_mash &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec, num_false_discoveries = num_false_discoveries_vec)
ggplot(num_discoveries_df_mash, aes(x = num_discoveries, y = (num_false_discoveries))) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Number of Discoveries&quot;, y = &quot;Number of False Discoveries&quot;) +
  theme_minimal() +
  ggtitle(&quot;MASH: Number of False Discoveries vs Number of Discoveries&quot;) +
  coord_cartesian(xlim = c(0, 300), ylim = c(0, 300))</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-24-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-24-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Compare MASH (fda_df_mash) and FASH (fda_df)</p>
<pre class="r"><code># Calculate true FDR for FASH
threshold_vec &lt;- seq(0.01, 0.99, by = 0.01)
fdr_vec_fash &lt;- numeric(length(threshold_vec))

for (i in 1:length(threshold_vec)) {
  num_discoveries_fash &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries_fash &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df$type == &quot;A&quot;)
  fdr_vec_fash[i] &lt;- num_false_discoveries_fash / num_discoveries_fash
}

# Create a data frame for plotting
fdr_df_fash &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_fash, method = &quot;FASH&quot;)

# Calculate true FDR for MASH
fdr_vec_mash &lt;- numeric(length(threshold_vec))

for (i in 1:length(threshold_vec)) {
  num_discoveries_mash &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries_mash &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df_mash$type == &quot;A&quot;)
  fdr_vec_mash[i] &lt;- num_false_discoveries_mash / num_discoveries_mash
}

# Create a data frame for plotting
fdr_df_mash &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_mash, method = &quot;MASH&quot;)

# Combine data for plotting
fdr_df_combined &lt;- rbind(fdr_df_fash, fdr_df_mash)

# Plot the nominal FDR vs true FDR for both methods
ggplot(fdr_df_combined, aes(x = threshold, y = true_fdr, color = method)) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Nominal False Discovery Rate&quot;, y = &quot;Actual False Discovery Rate&quot;) +
  theme_minimal() +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1)) +
  geom_hline(yintercept = sizeA/N, linetype = &quot;dashed&quot;, color = &quot;red&quot;) +
  ggtitle(&quot;Nominal FDR vs Actual FDR Curves for MASH and FASH&quot;)</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-25-1">
Past versions of unnamed-chunk-25-1.png
</button>
</p>
<div id="fig-unnamed-chunk-25-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-25-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-25-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Compare the number of false discoveries against the number of total
discoveries for both methods:</p>
<pre class="r"><code>threshold_vec &lt;- seq(0.01, 0.99, by = 0.01)
num_discoveries_vec_fash &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec_fash &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec_fash[i] &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries_vec_fash[i] &lt;- sum(fda_df$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df$type == &quot;A&quot;)
}
num_discoveries_df_fash &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec_fash, num_false_discoveries = num_false_discoveries_vec_fash, method = &quot;FASH&quot;)

num_discoveries_vec_mash &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec_mash &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec_mash[i] &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i])
  num_false_discoveries_vec_mash[i] &lt;- sum(fda_df_mash$cumulative_fda &lt;= threshold_vec[i] &amp; fda_df_mash$type == &quot;A&quot;)
}

num_discoveries_df_mash &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec_mash, num_false_discoveries = num_false_discoveries_vec_mash, method = &quot;MASH&quot;)

num_discoveries_df_combined &lt;- rbind(num_discoveries_df_fash, num_discoveries_df_mash)

ggplot(num_discoveries_df_combined, aes(x = num_discoveries, y = num_false_discoveries, color = method)) +
  geom_line() +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Number of Discoveries&quot;, y = &quot;Number of False Discoveries&quot;) +
  theme_minimal() +
  ggtitle(&quot;Number of False Discoveries vs Number of Discoveries for MASH and FASH&quot;) +
  coord_cartesian(xlim = c(0, 300), ylim = c(0, 300))</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-26-1">
Past versions of unnamed-chunk-26-1.png
</button>
</p>
<div id="fig-unnamed-chunk-26-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-26-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/9e1131d55208c45512bc391146047aa3a355c8f8/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-26-1.png" target="_blank">9e1131d</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>To compare their rankings of eQTLs based on the cumulative FDA, we
can compare the rank distribution of each type of eQTLs (A,B,C) based on
the two methods:</p>
<pre class="r"><code>fda_df_merge &lt;- rbind(fda_df %&gt;% mutate(method = &quot;FASH&quot;), fda_df_mash %&gt;% mutate(method = &quot;MASH&quot;))

ggplot(fda_df_merge, aes(x = type, y = rank, color = type)) +
  geom_boxplot() +
  labs(x = &quot;Type&quot;, y = &quot;Rank&quot;) +
  theme_minimal() +
  ggtitle(&quot;Rank Distribution: FASH vs MASH&quot;) +
  facet_wrap(~method)</code></pre>
<p><img src="figure/simulation_eQTL_identification.Rmd/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-27-1">
Past versions of unnamed-chunk-27-1.png
</button>
</p>
<div id="fig-unnamed-chunk-27-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/a53277262c0b48ae3703879a49f9f3a5c5fd6e4f/docs/figure/simulation_eQTL_identification.Rmd/unnamed-chunk-27-1.png" target="_blank">a532772</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-05-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><strong>Summary:</strong> Overall, FASH tends to make more
conservative decision than MASH in terms of the discovery of non-dynamic
eQTLs. The actual false discovery rate of FASH tends to be lower than
the nominal false discovery rate, while MASH tends to have a higher
actual false discovery rate than the nominal false discovery rate. This
kind of behavior is expected, as each mixture component of FASH is a
L-GP prior that shrinks toward the constant base model. In contrast, the
mixture components in MASH have different type of shrinkage tendencies.
which gains more modeling flexibility at the cost of less informative
shrinkage. Overall, MASH does better job at controlling an actual FDA
closer to the nominal level, but FASH does better job at keeping the
actual FDA lower than the nominal level.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Toronto
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] mashr_0.2.79    ashr_2.2-63     lubridate_1.9.3 forcats_1.0.0  
 [5] stringr_1.5.0   dplyr_1.1.3     purrr_1.0.2     readr_2.1.4    
 [9] tidyr_1.3.0     tibble_3.2.1    tidyverse_2.0.0 mixsqp_0.3-54  
[13] reshape2_1.4.4  ggplot2_3.5.1   Matrix_1.6-3    TMB_1.9.7      
[17] BayesGP_0.1.1   workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.4      xfun_0.40         bslib_0.5.1       processx_3.8.2   
 [5] lattice_0.21-8    callr_3.7.3       tzdb_0.4.0        vctrs_0.6.4      
 [9] tools_4.3.1       ps_1.7.5          generics_0.1.3    fansi_1.0.5      
[13] pkgconfig_2.0.3   SQUAREM_2021.1    assertthat_0.2.1  truncnorm_1.0-9  
[17] lifecycle_1.0.3   farver_2.1.1      compiler_4.3.1    git2r_0.33.0     
[21] munsell_0.5.0     getPass_0.2-4     httpuv_1.6.11     htmltools_0.5.6.1
[25] sass_0.4.7        yaml_2.3.7        later_1.3.1       pillar_1.9.0     
[29] jquerylib_0.1.4   whisker_0.4.1     rmeta_3.0         cachem_1.0.8     
[33] abind_1.4-5       tidyselect_1.2.0  digest_0.6.33     mvtnorm_1.2-3    
[37] stringi_1.7.12    labeling_0.4.3    rprojroot_2.0.3   fastmap_1.1.1    
[41] grid_4.3.1        invgamma_1.1      colorspace_2.1-0  cli_3.6.1        
[45] magrittr_2.0.3    utf8_1.2.3        withr_2.5.1       scales_1.3.0     
[49] promises_1.2.1    timechange_0.2.0  rmarkdown_2.25    httr_1.4.7       
[53] hms_1.1.3         evaluate_0.22     knitr_1.44        irlba_2.3.5.1    
[57] rlang_1.1.1       Rcpp_1.0.11       glue_1.6.2        rstudioapi_0.15.0
[61] jsonlite_1.8.7    R6_2.5.1          plyr_1.8.9        fs_1.6.3         </code></pre>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
