<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ziang Zhang" />

<meta name="date" content="2024-12-06" />

<title>Quantile regression for vQTLs</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">resultsummary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="simulation_pollution.html">Simulation 1</a>
</li>
<li>
  <a href="eQTL_sim.html">Simulation 2</a>
</li>
<li>
  <a href="simulation_vQTL.html">Simulation 3</a>
</li>
<li>
  <a href="covid_example.html">COVID-19 Example</a>
</li>
<li>
  <a href="expression.html">Gene Expression Example</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/AgueroZZ/resultsummary">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Quantile regression for vQTLs</h1>
<h4 class="author">Ziang Zhang</h4>
<h4 class="date">2024-12-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-12-13
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>FASHresultsummary/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240507code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240507)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240507code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240507)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreebbea11cca66aa4e8a8ebc73075c697475dc64428targetblankbbea11ca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/bbea11cca66aa4e8a8ebc73075c697475dc64428" target="_blank">bbea11c</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomAgueroZZFASHresultsummarytreebbea11cca66aa4e8a8ebc73075c697475dc64428targetblankbbea11ca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/AgueroZZ/FASHresultsummary/tree/bbea11cca66aa4e8a8ebc73075c697475dc64428" target="_blank">bbea11c</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    code/.DS_Store
    Ignored:    code/.Rhistory
    Ignored:    code/cpp/.DS_Store
    Ignored:    code/function/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/Iyer/.DS_Store
    Ignored:    data/expression_data/.DS_Store
    Ignored:    output/.DS_Store
    Ignored:    output/Iyer/.DS_Store
    Ignored:    output/example/.DS_Store
    Ignored:    output/example/figure/
    Ignored:    output/expression/.DS_Store
    Ignored:    output/simulation_pollution/.DS_Store
    Ignored:    output/simulation_pollution/figure/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/simulation_vQTL.Rmd</code>) and
HTML (<code>docs/simulation_vQTL.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/bbea11cca66aa4e8a8ebc73075c697475dc64428/analysis/simulation_vQTL.Rmd" target="_blank">bbea11c</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-13
</td>
<td>
workflowr::wflow_publish("analysis/simulation_vQTL.Rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/AgueroZZ/FASHresultsummary/622039578f498f54d56df63dd85253b4fa1bb413/docs/simulation_vQTL.html" target="_blank">6220395</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/61e22a457494dff85346ba5d693b2bfff4f93211/analysis/simulation_vQTL.Rmd" target="_blank">61e22a4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-12
</td>
<td>
workflowr::wflow_publish("analysis/simulation_vQTL.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2><em>Introduction</em></h2>
<p>In this example, we examined the performance of FASH in detecting
vQTLs based on the output from quantile regression. For a given SNP
<span class="math inline">\(G\)</span> and a continuous trait <span
class="math inline">\(Y\)</span>, the quantile regression model is given
by <span class="math display">\[
Q_Y(\tau|G = g) = g\beta(\tau),
\]</span> where <span class="math inline">\(Q_Y(\tau|G = g)\)</span> is
the <span class="math inline">\(\tau\)</span>-th quantile of <span
class="math inline">\(Y\)</span> given <span class="math inline">\(G =
g\)</span>, and <span class="math inline">\(\beta(\tau)\)</span> is the
effect size of <span class="math inline">\(G\)</span> on the <span
class="math inline">\(\tau\)</span>-th quantile of <span
class="math inline">\(Y\)</span>.</p>
<p>In <a
href="https://www.pnas.org/doi/epub/10.1073/pnas.2212959119">Miao et
al., 2022</a>, the authors proposed an approach called QUAIL, which uses
quantile regression to detect vQTLs. Specifically, if the SNP <span
class="math inline">\(G\)</span> has effect on the variance of <span
class="math inline">\(Y\)</span>, then its quantile effect <span
class="math inline">\(\beta(\tau)\)</span> will be different across
different quantiles <span class="math inline">\(\tau\)</span>. In other
words, the effect size <span class="math inline">\(\beta(\tau)\)</span>
will not be a constant function of <span
class="math inline">\(\tau\)</span>.</p>
<p>To test this hypothesis, the quantile effect <span
class="math inline">\(\beta(\tau)\)</span> is estimated for each
quantile <span class="math inline">\(\tau\)</span>, with estimates <span
class="math inline">\(\hat{\beta}(\tau)\)</span> and standard errors
<span class="math inline">\(SE(\hat{\beta}(\tau))\)</span>. QUAIL then
assesses the evidence of vQTLs by looking at the integrated effect <span
class="math inline">\(\beta_{QI} = \int_{0}^{0.5}[\beta(1-\tau) -
\beta(\tau)]d\tau\)</span>. If the null hypothesis of no vQTLs is true,
then <span class="math inline">\(\beta_{QI}\)</span> should be close to
zero.</p>
<p>The applications of quantile regression on biobank data can also be
found at <a
href="https://www.nature.com/articles/s42003-019-0387-5">Pozarickij et
al., 2019</a> and <a
href="https://www.nature.com/articles/s41467-024-50726-x">Wang et al.,
2024</a>, where the detected vQTLs were then linked to evidence of GxE
or GxG interactions.</p>
</div>
<div id="using-fash" class="section level2">
<h2><em>Using FASH</em></h2>
<p>Now, we will try to approach this problem using the FASH method.
Specifically, for each of the <span class="math inline">\(N\)</span>
SNPs <span class="math inline">\(\{G_1, ..., G_N\}\)</span>, we compute
their quantile regression effects <span
class="math inline">\(\{\hat{\beta}_i(\tau_j)\}\)</span> and standard
errors <span
class="math inline">\(\{SE(\hat{\beta}_i(\tau_j))\}\)</span>, at a set
of quantiles <span class="math inline">\(\{\tau_1, ...,
\tau_J\}\)</span>. We then assume the following model for the estimated
quantile effects: <span class="math display">\[
\hat{\beta}_i(\tau_j) \overset{ind}{\sim} N(\beta_i(\tau_j),
SE(\hat{\beta}_i(\tau_j))^2),
\]</span> where <span
class="math inline">\(\beta_i(\tau)|\pi_0,...,\pi_K \overset{iid}{\sim}
\sum_{k=0}^{K} \pi_k\text{IWP}_1(\sigma_k)\)</span>. For a function
<span class="math inline">\(g \sim \text{IWP}_1(\sigma)\)</span>, the
prior is given by: <span class="math display">\[
\frac{\partial}{\partial \tau}\beta(\tau) = \sigma_k\xi(\tau),
\]</span> where <span class="math inline">\(\beta(0)\)</span> is
assigned a diffuse prior and <span
class="math inline">\(\xi(\tau)\)</span> is the standard Gaussian white
noise.</p>
<p>The weight <span class="math inline">\(\pi_0\)</span> corresponds to
the base model where <span class="math inline">\(\sigma_0 = 0\)</span>
and <span class="math inline">\(\beta(\tau) =
\text{span}\{1\}\)</span>.</p>
</div>
<div id="simulation-setup" class="section level2">
<h2><em>Simulation Setup</em></h2>
<p>To assess the performance of FASH, we simulate <span
class="math inline">\(N = 10000\)</span> datasets of G and Y.</p>
<ul>
<li>In the first <span class="math inline">\(N_A = 1000\)</span>
datasets, the G is assumed to affect Y on its variance level
(vQTLs).</li>
<li>In the next <span class="math inline">\(N_B = 1000\)</span>
datasets, the G is assumed to affect Y on its mean lev (not vQTLs).</li>
<li>In the next <span class="math inline">\(N_C = 1000\)</span>
datasets, the G is assumed to affect Y on its mean and variance levels
(vQTLs).</li>
<li>For the other <span class="math inline">\(N_D = 7000\)</span>
datasets, the G is assumed to have no effect on Y (not vQTLs).</li>
</ul>
<p>The true effect, if exists, is assumed to be normally distributed
with mean 0 and standard deviation <code>beta_sd</code>.</p>
<p>Once the data is simulated, we can perform quantile regression on
each dataset to obtain the estimates <span
class="math inline">\(\hat{\beta}(\tau)\)</span> and standard errors
<span class="math inline">\(SE(\hat{\beta}(\tau))\)</span>.</p>
<p>Here we can visualize one example of the quantile regression
results.</p>
<p><img src="figure/simulation_vQTL.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-2-1">
Past versions of unnamed-chunk-2-1.png
</button>
</p>
<div id="fig-unnamed-chunk-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/622039578f498f54d56df63dd85253b4fa1bb413/docs/figure/simulation_vQTL.Rmd/unnamed-chunk-2-1.png" target="_blank">6220395</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-12
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now, let’s do the simulation with FASH:</p>
<pre class="r"><code># Parameters
set.seed(123)  # For reproducibility
n &lt;- 20000      # Number of observations per dataset
maf &lt;- 0.3     # Minor allele frequency
beta_sd &lt;- 0.3   # Effect size standard deviation
sigma &lt;- 1     # Standard deviation
tau_vec &lt;- seq(0.1, 0.9, by = 0.1)  # Quantiles for regression
N_A &lt;- 500; N_B &lt;- 500; N_C &lt;- 500; N_D &lt;- 3500  # Number of datasets for each case
num_datasets &lt;- N_A + N_B + N_C + N_D # Total number of datasets

# Initialize matrices to store beta estimates and standard errors
beta_matrix &lt;- matrix(NA, nrow = num_datasets, ncol = length(tau_vec))
se_matrix &lt;- matrix(NA, nrow = num_datasets, ncol = length(tau_vec))

# Generate datasets and perform quantile regression
for (i in 1:num_datasets) {
  if (i &lt;= N_A) {
    case &lt;- &quot;variance&quot;
  } else if (i &lt;= (N_A + N_B)) {
    case &lt;- &quot;mean&quot;
  } else if (i &lt;= (N_A + N_B + N_C)) {
    case &lt;- &quot;both&quot;
  } else {
    case &lt;- &quot;none&quot;
  }
  
  # Simulate the data
  data &lt;- simulate_data(n = n, maf = maf, beta_sd = beta_sd, sigma = sigma, case = case)
  
  # Perform quantile regression
  result &lt;- perform_rq(data, tau_vec = tau_vec, bootstrap = F)
  
  # print how many datasets have been simulated in a message
  if(i %% 100 == 0){
    print(paste0(i, &quot; datasets have been simulated.&quot;))
  }
  
  # Store beta estimates and standard errors
  beta_matrix[i, ] &lt;- result$beta
  se_matrix[i, ] &lt;- result$se
}

# Output matrices
print(dim(beta_matrix))  # Check dimensions
print(dim(se_matrix))    # Check dimensions</code></pre>
<pre class="r"><code>p_vec &lt;- 1
psd_vec &lt;- seq(0, 0.5, length = 50)

# Using the FEM:
L_vecs &lt;- mclapply(datasets, function(dataset) {
  compute_log_likelihood_ospline_seq2(
    x = dataset$tau,
    y = dataset$beta,
    p = p_vec,
    num_knots = 10,
    psd_iwp_vector = psd_vec,
    pred_step = 1/10,
    betaprec = 1e-6,
    sd_gaussian = dataset$se
  )
}, mc.cores = 1)
L_matrix &lt;- do.call(rbind, L_vecs)

save(L_matrix, file = paste0(result_dir, &quot;/L_matrix.rda&quot;))</code></pre>
<p>Learning the prior weights by EB: (with possible penalty term)</p>
<pre class="r"><code>prior_null_lambda = 1
if(prior_null_lambda &gt; 1){
  prior_null &lt;- matrix(0, nrow = (prior_null_lambda-1), ncol = ncol(L_matrix))
  prior_null[,1] &lt;- 1
  L_matrix_original &lt;- rbind(exp(L_matrix), prior_null)
  fit.sqp &lt;- mixsqp(L = L_matrix_original, log = FALSE, control = list(tol.svd = 0))
}else{
  fit.sqp &lt;- mixsqp(L = L_matrix, log = TRUE, control = list(tol.svd = 0))
}</code></pre>
<pre><code>Running mix-SQP algorithm 0.3-54 on 5000 x 50 matrix
convergence tol. (SQP):     1.0e-08
conv. tol. (active-set):    1.0e-10
zero threshold (solution):  1.0e-08
zero thresh. (search dir.): 1.0e-14
l.s. sufficient decrease:   1.0e-02
step size reduction factor: 7.5e-01
minimum step size:          1.0e-08
max. iter (SQP):            1000
max. iter (active-set):     20
number of EM iterations:    10
iter        objective max(rdual) nnz stepsize max.diff nqp nls
   1 +7.078216772e-01  -- EM --   50 1.00e+00 1.55e-01  --  --
   2 +6.217476781e-01  -- EM --   48 1.00e+00 6.83e-02  --  --
   3 +6.058465200e-01  -- EM --   35 1.00e+00 3.76e-02  --  --
   4 +5.993127379e-01  -- EM --   30 1.00e+00 2.52e-02  --  --
   5 +5.952752725e-01  -- EM --   27 1.00e+00 1.91e-02  --  --
   6 +5.924442531e-01  -- EM --   25 1.00e+00 1.56e-02  --  --
   7 +5.903640093e-01  -- EM --   24 1.00e+00 1.33e-02  --  --
   8 +5.887899615e-01  -- EM --   23 1.00e+00 1.17e-02  --  --
   9 +5.875703750e-01  -- EM --   22 1.00e+00 1.03e-02  --  --
  10 +5.866060570e-01  -- EM --   21 1.00e+00 9.27e-03  --  --
   1 +5.858300717e-01 +2.967e-02  21  ------   ------   --  --
   2 +5.796489016e-01 +1.459e-02   6 1.00e+00 3.66e-01  20   1
   3 +5.794165156e-01 +1.682e-03   6 1.00e+00 5.17e-02   4   1
   4 +5.794147284e-01 +4.066e-05   5 1.00e+00 2.23e-03   3   1
   5 +5.794147276e-01 -1.778e-08   5 1.00e+00 4.96e-05   2   1
Optimization took 0.06 seconds.
Convergence criteria met---optimal solution found.</code></pre>
<pre class="r"><code>numiter &lt;- nrow(fit.sqp$progress)
prior_weight &lt;- data.frame(p = rep(p_vec, each = length(psd_vec)), psd_iwp = psd_vec, prior_weight = fit.sqp$x)
prior_weight$prior_weight[1]</code></pre>
<pre><code>[1] 0.7259555</code></pre>
<pre class="r"><code>posterior_matrix &lt;- matrix(0, nrow = nrow(L_matrix), ncol = ncol(L_matrix))
for(i in 1:nrow(L_matrix)){
  posterior_matrix[i,] &lt;- exp(L_matrix[i,] - max(L_matrix[i,]) + log(fit.sqp$x))
  posterior_matrix[i,] &lt;- posterior_matrix[i,]/sum(posterior_matrix[i,])
}
colnames(posterior_matrix) &lt;- paste0(p_vec,&quot;_&quot;,psd_vec)
posterior_weights_matrix &lt;- posterior_matrix</code></pre>
<p>Visualize the posterior probability for each SNP being in each of the
<span class="math inline">\(K\)</span> classes:</p>
<pre class="r"><code>posterior_weights_df &lt;- as.data.frame(posterior_weights_matrix)
posterior_weights_df$id &lt;- 1:nrow(posterior_weights_df)
melted_data &lt;- melt(posterior_weights_df, id.vars = &quot;id&quot;)
melted_data$variable2 &lt;- sub(&quot;_.*&quot;, &quot;&quot;, melted_data$variable)
melted_data$variable3 &lt;- (round(as.numeric(sub(&quot;*._&quot;, &quot;&quot;, melted_data$variable)), 3))

ggplot(melted_data, aes(x = factor(id, levels = posterior_weights_df$id), y = value, fill = variable3)) +
  geom_bar(stat = &quot;identity&quot;) +
  labs(x = &quot;SNPs&quot;, y = &quot;Weight&quot;, fill = &quot;PSD&quot;) +
  ggtitle(&quot;Structure Plot of Posterior Weights&quot;) +
  coord_flip() +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;blue&quot;) +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    panel.background = element_rect(fill = &quot;white&quot;, colour = &quot;grey&quot;),
    plot.background = element_rect(fill = &quot;white&quot;, colour = NA)
  )</code></pre>
<p><img src="figure/simulation_vQTL.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-9-1">
Past versions of unnamed-chunk-9-1.png
</button>
</p>
<div id="fig-unnamed-chunk-9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/622039578f498f54d56df63dd85253b4fa1bb413/docs/figure/simulation_vQTL.Rmd/unnamed-chunk-9-1.png" target="_blank">6220395</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-12
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="fdr-computed-using-fash" class="section level4">
<h4><em>FDR Computed Using FASH</em></h4>
<p>Compute the local false discovery rate (lfdr):</p>
<pre class="r"><code>set.seed(123)
lfdr &lt;- posterior_weights_matrix[,1]
fdr_df &lt;- data.frame(eQTL = 1:length(lfdr), fdr = lfdr, type = rep(c(&quot;variance&quot;, &quot;mean&quot;, &quot;both&quot;, &quot;none&quot;), times = c(N_A, N_B, N_C, N_D)))
fdr_df &lt;- fdr_df[order(fdr_df$fdr), ] # ordering it
fdr_df$cumulative_fdr &lt;- cumsum(fdr_df$fdr)/seq_along(fdr_df$fdr)
fdr_df$rank &lt;- 1:length(lfdr)
fdr_df$max_fdr &lt;- cummax(fdr_df$fdr)</code></pre>
</div>
<div id="fdr-computed-using-mash" class="section level4">
<h4><em>FDR Computed Using MASH</em></h4>
<p>This problem could also be easily solved by MASH, by extracting its
components corresponding to <code>equal_effects</code>. Let’s do it and
compare the result:</p>
<pre class="r"><code>fitted_datasets_mash &lt;- list()

# Produce a huge data-matrix, the i-th row being dataset[[i]]$beta
all_data_matrix &lt;- do.call(rbind, lapply(datasets, function(x) x$beta))
SE_matrix &lt;- do.call(rbind, lapply(datasets, function(x) x$se))

# now use mashr:
mash_data &lt;-  mashr::mash_set_data(all_data_matrix, SE_matrix)
m.1by1 = mashr::mash_1by1(mash_data)
strong = mashr::get_significant_results(m.1by1, 0.05)
# keep the top 10%
strong &lt;- strong[1:round(0.1*length(strong))]
U.pca = mashr::cov_pca(mash_data, 5, subset = strong)
U.ed = cov_ed(mash_data, U.pca, subset=strong)
U.c = cov_canonical(mash_data)
m   = mash(mash_data, c(U.c,U.ed))
# m   = mashr::mash(mash_data, U.c)
save(m, file = paste0(result_dir, &quot;/mash_result.rda&quot;))</code></pre>
<pre class="r"><code>load(paste0(result_dir, &quot;/mash_result.rda&quot;))
mash_post &lt;- m$posterior_weights
## extract the colnames start with &quot;equal_effects&quot;
lfdr_mash &lt;- mash_post[, c(1, grep(&quot;equal_effects&quot;, colnames(mash_post)))]
# sum each row
lfdr_mash &lt;- rowSums(lfdr_mash)

fdr_df_mash &lt;- data.frame(SNPs = 1:length(lfdr_mash), fdr = lfdr_mash, type = rep(c(&quot;variance&quot;, &quot;mean&quot;, &quot;both&quot;, &quot;none&quot;), times = c(N_A, N_B, N_C, N_D)))

fdr_df_mash &lt;- fdr_df_mash[order(fdr_df_mash$fdr), ] # ordering it
fdr_df_mash$cumulative_fdr &lt;- cumsum(fdr_df_mash$fdr)/seq_along(fdr_df_mash$fdr)
fdr_df_mash$rank &lt;- 1:length(lfdr_mash)</code></pre>
</div>
<div id="comparison" class="section level4">
<h4><em>Comparison</em></h4>
<pre class="r"><code># Calculate true FDR for FASH and MASH
threshold_vec &lt;- seq(0, 1, by = 0.005)[-1]
fdr_vec_fash &lt;- numeric(length(threshold_vec))
fdr_vec_mash &lt;- numeric(length(threshold_vec))

for (i in 1:length(threshold_vec)) {
  num_discoveries_fash &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries_fash &lt;- sum(fdr_df$cumulative_fdr &lt;= threshold_vec[i] &amp; (fdr_df$type == &quot;none&quot; | fdr_df$type == &quot;mean&quot;))
  fdr_vec_fash[i] &lt;- num_false_discoveries_fash / num_discoveries_fash
  
  num_discoveries_mash &lt;- sum(fdr_df_mash$cumulative_fdr &lt;= threshold_vec[i])
  num_false_discoveries_mash &lt;- sum(fdr_df_mash$cumulative_fdr &lt;= threshold_vec[i] &amp; (fdr_df_mash$type == &quot;none&quot; | fdr_df_mash$type == &quot;mean&quot;))
  fdr_vec_mash[i] &lt;- num_false_discoveries_mash / num_discoveries_mash
  
}

# Create a data frame for plotting
fdr_df_fash &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_fash, method = &quot;FASH&quot;)
fdr_df_fash_for_plotting &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_fash, method = &quot;FASH&quot;)
fd_df_mash &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_mash, method = &quot;MASH&quot;)
fdr_df_mash_for_plotting &lt;- data.frame(threshold = threshold_vec, true_fdr = fdr_vec_mash, method = &quot;MASH&quot;)

# Combine data for plotting
fdr_df_combined &lt;- rbind(fdr_df_fash_for_plotting, fdr_df_mash_for_plotting)

# Plot the nominal FDR vs true FDR for both methods
ggplot(fdr_df_combined, aes(x = threshold, y = true_fdr, color = method)) +
  geom_line() +
  # geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Nominal False Discovery Rate&quot;, y = &quot;Actual False Discovery Rate&quot;) +
  theme_minimal() +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1)) +
  geom_hline(yintercept = (N_A + N_D)/num_datasets, linetype = &quot;dashed&quot;, color = &quot;red&quot;) +
  ggtitle(&quot;Nominal FDR vs Actual FDR Curves for FASH and MASH&quot;)</code></pre>
<p><img src="figure/simulation_vQTL.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/622039578f498f54d56df63dd85253b4fa1bb413/docs/figure/simulation_vQTL.Rmd/unnamed-chunk-13-1.png" target="_blank">6220395</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-12
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>threshold_vec &lt;- seq(0, 1, by = 0.001)
num_discoveries_vec_fash &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec_fash &lt;- numeric(length(threshold_vec))
num_discoveries_vec_mash &lt;- numeric(length(threshold_vec))
num_false_discoveries_vec_mash &lt;- numeric(length(threshold_vec))
for (i in 1:length(threshold_vec)) {
  num_discoveries_vec_fash[i] &lt;- sum(fdr_df$cumulative_fdr &lt; threshold_vec[i])
  num_false_discoveries_vec_fash[i] &lt;- sum(fdr_df$cumulative_fdr &lt; threshold_vec[i] &amp; (fdr_df$type == &quot;none&quot; | fdr_df$type == &quot;mean&quot;))
  
  num_discoveries_vec_mash[i] &lt;- sum(fdr_df_mash$cumulative_fdr &lt; threshold_vec[i])
  num_false_discoveries_vec_mash[i] &lt;- sum(fdr_df_mash$cumulative_fdr &lt; threshold_vec[i] &amp; (fdr_df_mash$type == &quot;none&quot; | fdr_df_mash$type == &quot;mean&quot;))
}
num_discoveries_df_fash &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec_fash, num_false_discoveries = num_false_discoveries_vec_fash, method = &quot;FASH&quot;)
num_discoveries_df_mash &lt;- data.frame(threshold = threshold_vec, num_discoveries = num_discoveries_vec_mash, num_false_discoveries = num_false_discoveries_vec_mash, method = &quot;MASH&quot;)

# combined for plotting
num_discoveries_df_combined &lt;- rbind(num_discoveries_df_fash, num_discoveries_df_mash)

ggplot(num_discoveries_df_combined, aes(x = num_discoveries, y = num_false_discoveries, color = method)) +
  geom_line(aes(linetype = method)) +
  geom_point(size = 0.1) +
  geom_abline(intercept = 0, slope = 1, linetype = &quot;dashed&quot;, color = &quot;purple&quot;) +
  labs(x = &quot;Number of Discoveries&quot;, y = &quot;Number of False Discoveries&quot;) +
  theme_minimal() +
  geom_vline(xintercept = (N_A + N_C), linetype = &quot;dashed&quot;, color = &quot;red&quot;) +
  ggtitle(&quot;Number of False Discoveries vs Number of Discoveries for FASH&quot;) +
  coord_cartesian(xlim = c(0, num_datasets), ylim = c(0, num_datasets))</code></pre>
<p><img src="figure/simulation_vQTL.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/622039578f498f54d56df63dd85253b4fa1bb413/docs/figure/simulation_vQTL.Rmd/unnamed-chunk-14-1.png" target="_blank">6220395</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-12
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code># plot power versus FDR
power_vec_fash &lt;- (num_discoveries_df_fash$num_discoveries - num_discoveries_df_fash$num_false_discoveries) / (N_A + N_C)
power_df_fash &lt;- data.frame(threshold = num_discoveries_df_fash$threshold, power = power_vec_fash, method = &quot;FASH&quot;)
power_vec_mash &lt;- (num_discoveries_df_mash$num_discoveries - num_discoveries_df_mash$num_false_discoveries) / (N_A + N_C)
power_df_mash &lt;- data.frame(threshold = num_discoveries_df_mash$threshold, power = power_vec_mash, method = &quot;MASH&quot;)

power_df_combined &lt;- rbind(power_df_fash, power_df_mash)

ggplot(power_df_combined, aes(x = threshold, y = power, color = method)) +
  geom_line() +
  geom_point(size = 0.1) +
  labs(x = &quot;Nominal False Discovery Rate&quot;, y = &quot;Power&quot;) +
  theme_minimal() +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1)) +
  ggtitle(&quot;Power vs Nominal FDR Curves for FASH and MASH&quot;)</code></pre>
<p><img src="figure/simulation_vQTL.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-1">
Past versions of unnamed-chunk-15-1.png
</button>
</p>
<div id="fig-unnamed-chunk-15-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/AgueroZZ/FASHresultsummary/blob/622039578f498f54d56df63dd85253b4fa1bb413/docs/figure/simulation_vQTL.Rmd/unnamed-chunk-15-1.png" target="_blank">6220395</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2024-12-12
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] quantreg_5.98   SparseM_1.84-2  mashr_0.2.79    ashr_2.2-66    
 [5] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [9] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   
[13] tidyverse_2.0.0 mixsqp_0.3-54   reshape2_1.4.4  ggplot2_3.5.1  
[17] Matrix_1.6-4    TMB_1.9.15      BayesGP_0.1.3   workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.6       xfun_0.48          bslib_0.8.0        processx_3.8.4    
 [5] lattice_0.22-6     callr_3.7.6        tzdb_0.4.0         vctrs_0.6.5       
 [9] tools_4.3.1        ps_1.8.0           generics_0.1.3     fansi_1.0.6       
[13] highr_0.11         pkgconfig_2.0.3    SQUAREM_2021.1     assertthat_0.2.1  
[17] truncnorm_1.0-9    lifecycle_1.0.4    farver_2.1.2       compiler_4.3.1    
[21] git2r_0.33.0       MatrixModels_0.5-3 munsell_0.5.1      getPass_0.2-4     
[25] httpuv_1.6.15      htmltools_0.5.8.1  sass_0.4.9         yaml_2.3.10       
[29] later_1.3.2        pillar_1.9.0       jquerylib_0.1.4    whisker_0.4.1     
[33] MASS_7.3-60        rmeta_3.0          cachem_1.1.0       abind_1.4-8       
[37] tidyselect_1.2.1   digest_0.6.37      mvtnorm_1.3-1      stringi_1.8.4     
[41] labeling_0.4.3     rprojroot_2.0.4    fastmap_1.2.0      grid_4.3.1        
[45] invgamma_1.1       colorspace_2.1-1   cli_3.6.3          magrittr_2.0.3    
[49] survival_3.7-0     utf8_1.2.4         withr_3.0.2        scales_1.3.0      
[53] promises_1.3.0     timechange_0.3.0   rmarkdown_2.28     httr_1.4.7        
[57] hms_1.1.3          evaluate_1.0.1     knitr_1.48         irlba_2.3.5.1     
[61] rlang_1.1.4        Rcpp_1.0.13-1      glue_1.8.0         rstudioapi_0.16.0 
[65] jsonlite_1.8.9     R6_2.5.1           plyr_1.8.9         fs_1.6.4          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
